// @visactor/vchart/cjs/animation/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.shouldMarkDoMorph = exports.userAnimationConfig = exports.animationConfig = exports.DEFAULT_ANIMATION_CONFIG = exports.registerScaleInOutAnimation = exports.registerArcAnimation = exports.registerRectAnimation = exports.registerPolygonAnimation = exports.registerSequentialAnimate = exports.registerStateTransition = exports.registerCustomAnimate = exports.registerVRenderAnimate = void 0;

var vrender_animate_1 = require("@visactor/vrender-animate");

Object.defineProperty(exports, "registerVRenderAnimate", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.registerAnimate;
    }
}), Object.defineProperty(exports, "registerCustomAnimate", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.registerCustomAnimate;
    }
});

var state_transition_1 = require("./state-transition");

Object.defineProperty(exports, "registerStateTransition", {
    enumerable: !0,
    get: function() {
        return state_transition_1.registerStateTransition;
    }
});

var sequential_animate_1 = require("./sequential-animate");

Object.defineProperty(exports, "registerSequentialAnimate", {
    enumerable: !0,
    get: function() {
        return sequential_animate_1.registerSequentialAnimate;
    }
});

var config_1 = require("./config");

Object.defineProperty(exports, "registerPolygonAnimation", {
    enumerable: !0,
    get: function() {
        return config_1.registerPolygonAnimation;
    }
}), Object.defineProperty(exports, "registerRectAnimation", {
    enumerable: !0,
    get: function() {
        return config_1.registerRectAnimation;
    }
}), Object.defineProperty(exports, "registerArcAnimation", {
    enumerable: !0,
    get: function() {
        return config_1.registerArcAnimation;
    }
}), Object.defineProperty(exports, "registerScaleInOutAnimation", {
    enumerable: !0,
    get: function() {
        return config_1.registerScaleInOutAnimation;
    }
}), Object.defineProperty(exports, "DEFAULT_ANIMATION_CONFIG", {
    enumerable: !0,
    get: function() {
        return config_1.DEFAULT_ANIMATION_CONFIG;
    }
});

var utils_1 = require("./utils");

Object.defineProperty(exports, "animationConfig", {
    enumerable: !0,
    get: function() {
        return utils_1.animationConfig;
    }
}), Object.defineProperty(exports, "userAnimationConfig", {
    enumerable: !0,
    get: function() {
        return utils_1.userAnimationConfig;
    }
}), Object.defineProperty(exports, "shouldMarkDoMorph", {
    enumerable: !0,
    get: function() {
        return utils_1.shouldMarkDoMorph;
    }
});
