// @visactor/vrender-animate/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.AnimationStateStore = exports.AnimationStateManager = exports.transitionRegistry = exports.AnimationTransitionRegistry = exports.AnimateExecutor = exports.RotateBySphereAnimate = exports.GroupFadeOut = exports.GroupFadeIn = exports.TagPointsUpdate = exports.ClipDirectionAnimate = exports.ClipRadiusAnimate = exports.ClipAngleAnimate = exports.ClipGraphicAnimate = exports.InputText = exports.morphPath = exports.multiToOneMorph = exports.oneToMultiMorph = exports.MultiToOneMorphingPath = exports.MorphingPath = exports.IncreaseCount = exports.ComponentAnimator = exports.AStageAnimate = exports.AComponentAnimate = exports.ACustomAnimate = exports.registerAnimate = exports.AnimateStep = exports.DefaultTicker = exports.ManualTicker = exports.DefaultTimeline = exports.Animate = void 0;

var animate_1 = require("./animate");

Object.defineProperty(exports, "Animate", {
    enumerable: !0,
    get: function() {
        return animate_1.Animate;
    }
});

var timeline_1 = require("./timeline");

Object.defineProperty(exports, "DefaultTimeline", {
    enumerable: !0,
    get: function() {
        return timeline_1.DefaultTimeline;
    }
});

var manual_ticker_1 = require("./ticker/manual-ticker");

Object.defineProperty(exports, "ManualTicker", {
    enumerable: !0,
    get: function() {
        return manual_ticker_1.ManualTicker;
    }
});

var default_ticker_1 = require("./ticker/default-ticker");

Object.defineProperty(exports, "DefaultTicker", {
    enumerable: !0,
    get: function() {
        return default_ticker_1.DefaultTicker;
    }
});

var step_1 = require("./step");

Object.defineProperty(exports, "AnimateStep", {
    enumerable: !0,
    get: function() {
        return step_1.Step;
    }
}), __exportStar(require("./utils/easing-func"), exports), __exportStar(require("./utils/easing"), exports);

var register_1 = require("./register");

Object.defineProperty(exports, "registerAnimate", {
    enumerable: !0,
    get: function() {
        return register_1.registerAnimate;
    }
});

var custom_animate_1 = require("./custom/custom-animate");

Object.defineProperty(exports, "ACustomAnimate", {
    enumerable: !0,
    get: function() {
        return custom_animate_1.ACustomAnimate;
    }
}), Object.defineProperty(exports, "AComponentAnimate", {
    enumerable: !0,
    get: function() {
        return custom_animate_1.AComponentAnimate;
    }
}), Object.defineProperty(exports, "AStageAnimate", {
    enumerable: !0,
    get: function() {
        return custom_animate_1.AStageAnimate;
    }
});

var component_animator_1 = require("./component/component-animator");

Object.defineProperty(exports, "ComponentAnimator", {
    enumerable: !0,
    get: function() {
        return component_animator_1.ComponentAnimator;
    }
});

var number_1 = require("./custom/number");

Object.defineProperty(exports, "IncreaseCount", {
    enumerable: !0,
    get: function() {
        return number_1.IncreaseCount;
    }
});

var morphing_1 = require("./custom/morphing");

Object.defineProperty(exports, "MorphingPath", {
    enumerable: !0,
    get: function() {
        return morphing_1.MorphingPath;
    }
}), Object.defineProperty(exports, "MultiToOneMorphingPath", {
    enumerable: !0,
    get: function() {
        return morphing_1.MultiToOneMorphingPath;
    }
}), Object.defineProperty(exports, "oneToMultiMorph", {
    enumerable: !0,
    get: function() {
        return morphing_1.oneToMultiMorph;
    }
}), Object.defineProperty(exports, "multiToOneMorph", {
    enumerable: !0,
    get: function() {
        return morphing_1.multiToOneMorph;
    }
}), Object.defineProperty(exports, "morphPath", {
    enumerable: !0,
    get: function() {
        return morphing_1.morphPath;
    }
});

var input_text_1 = require("./custom/input-text");

Object.defineProperty(exports, "InputText", {
    enumerable: !0,
    get: function() {
        return input_text_1.InputText;
    }
});

var clip_graphic_1 = require("./custom/clip-graphic");

Object.defineProperty(exports, "ClipGraphicAnimate", {
    enumerable: !0,
    get: function() {
        return clip_graphic_1.ClipGraphicAnimate;
    }
}), Object.defineProperty(exports, "ClipAngleAnimate", {
    enumerable: !0,
    get: function() {
        return clip_graphic_1.ClipAngleAnimate;
    }
}), Object.defineProperty(exports, "ClipRadiusAnimate", {
    enumerable: !0,
    get: function() {
        return clip_graphic_1.ClipRadiusAnimate;
    }
}), Object.defineProperty(exports, "ClipDirectionAnimate", {
    enumerable: !0,
    get: function() {
        return clip_graphic_1.ClipDirectionAnimate;
    }
});

var tag_points_1 = require("./custom/tag-points");

Object.defineProperty(exports, "TagPointsUpdate", {
    enumerable: !0,
    get: function() {
        return tag_points_1.TagPointsUpdate;
    }
});

var groupFade_1 = require("./custom/groupFade");

Object.defineProperty(exports, "GroupFadeIn", {
    enumerable: !0,
    get: function() {
        return groupFade_1.GroupFadeIn;
    }
}), Object.defineProperty(exports, "GroupFadeOut", {
    enumerable: !0,
    get: function() {
        return groupFade_1.GroupFadeOut;
    }
});

var sphere_1 = require("./custom/sphere");

Object.defineProperty(exports, "RotateBySphereAnimate", {
    enumerable: !0,
    get: function() {
        return sphere_1.RotateBySphereAnimate;
    }
});

var animate_executor_1 = require("./executor/animate-executor");

Object.defineProperty(exports, "AnimateExecutor", {
    enumerable: !0,
    get: function() {
        return animate_executor_1.AnimateExecutor;
    }
}), __exportStar(require("./custom/register"), exports), __exportStar(require("./state"), exports);

var animation_states_registry_1 = require("./state/animation-states-registry");

Object.defineProperty(exports, "AnimationTransitionRegistry", {
    enumerable: !0,
    get: function() {
        return animation_states_registry_1.AnimationTransitionRegistry;
    }
});

var animation_states_registry_2 = require("./state/animation-states-registry");

Object.defineProperty(exports, "transitionRegistry", {
    enumerable: !0,
    get: function() {
        return animation_states_registry_2.transitionRegistry;
    }
});

var animation_state_1 = require("./state/animation-state");

Object.defineProperty(exports, "AnimationStateManager", {
    enumerable: !0,
    get: function() {
        return animation_state_1.AnimationStateManager;
    }
});

var animation_state_2 = require("./state/animation-state");

Object.defineProperty(exports, "AnimationStateStore", {
    enumerable: !0,
    get: function() {
        return animation_state_2.AnimationStateStore;
    }
}), __exportStar(require("./component"), exports);
