// @visactor/vchart/cjs/chart/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerLinearProgressChart = exports.registerHistogramChart = exports.registerHeatmapChart = exports.registerGaugeChart = exports.registerFunnelChart = exports.registerCommonChart = exports.registerCircularProgressChart = exports.registerCirclePackingChart = exports.registerBoxplotChart = exports.registerBarChart = exports.registerAreaChart = exports.registerLineChart = exports.MosaicChart = exports.BaseChart = exports.VennChart = exports.TreemapChart = exports.WordCloudChart = exports.BaseWordCloudChart = exports.BaseWordCloudChartSpecTransformer = exports.LiquidChart = exports.CorrelationChart = exports.WaterfallChart = exports.SequenceChart = exports.SunburstChart = exports.ScatterChart = exports.SankeyChart = exports.RoseChart = exports.RangeAreaChart = exports.RangeColumnChart = exports.RadarChart = exports.LinearProgressChart = exports.CircularProgressChart = exports.PieChart = exports.BasePieChart = exports.BasePieChartSpecTransformer = exports.MapChart = exports.LineChart = exports.HistogramChart = exports.BaseHistogramChart = exports.HistogramChartSpecTransformer = exports.HeatmapChart = exports.GaugeChart = exports.FunnelChart = exports.FunnelChartSpecTransformer = exports.CommonChart = exports.CirclePackingChart = exports.BoxPlotChart = exports.BarChartSpecTransformer = exports.BarChart = exports.AreaChart = void 0, 
exports.setDefaultCrosshairForCartesianChart = exports.registerMosaicChart = exports.registerVennChart = exports.registerWordCloudShapeChart = exports.registerLiquidChart = exports.registerCorrelationChart = exports.registerWordCloudChart = exports.registerWaterfallChart = exports.registerTreemapChart = exports.registerSunburstChart = exports.registerSequenceChart = exports.registerScatterChart = exports.registerSankeyChart = exports.registerRoseChart = exports.registerRangeColumnChart = exports.registerRangeAreaChart = exports.registerRadarChart = exports.registerPieChart = exports.registerMapChart = void 0;

const area_1 = require("./area");

Object.defineProperty(exports, "AreaChart", {
    enumerable: !0,
    get: function() {
        return area_1.AreaChart;
    }
}), Object.defineProperty(exports, "registerAreaChart", {
    enumerable: !0,
    get: function() {
        return area_1.registerAreaChart;
    }
});

const bar_1 = require("./bar");

Object.defineProperty(exports, "BarChart", {
    enumerable: !0,
    get: function() {
        return bar_1.BarChart;
    }
}), Object.defineProperty(exports, "registerBarChart", {
    enumerable: !0,
    get: function() {
        return bar_1.registerBarChart;
    }
}), Object.defineProperty(exports, "BarChartSpecTransformer", {
    enumerable: !0,
    get: function() {
        return bar_1.BarChartSpecTransformer;
    }
});

const box_plot_1 = require("./box-plot");

Object.defineProperty(exports, "BoxPlotChart", {
    enumerable: !0,
    get: function() {
        return box_plot_1.BoxPlotChart;
    }
}), Object.defineProperty(exports, "registerBoxplotChart", {
    enumerable: !0,
    get: function() {
        return box_plot_1.registerBoxplotChart;
    }
});

const circle_packing_1 = require("./circle-packing");

Object.defineProperty(exports, "CirclePackingChart", {
    enumerable: !0,
    get: function() {
        return circle_packing_1.CirclePackingChart;
    }
}), Object.defineProperty(exports, "registerCirclePackingChart", {
    enumerable: !0,
    get: function() {
        return circle_packing_1.registerCirclePackingChart;
    }
});

const common_1 = require("./common");

Object.defineProperty(exports, "CommonChart", {
    enumerable: !0,
    get: function() {
        return common_1.CommonChart;
    }
}), Object.defineProperty(exports, "registerCommonChart", {
    enumerable: !0,
    get: function() {
        return common_1.registerCommonChart;
    }
});

const funnel_1 = require("./funnel");

Object.defineProperty(exports, "FunnelChart", {
    enumerable: !0,
    get: function() {
        return funnel_1.FunnelChart;
    }
}), Object.defineProperty(exports, "registerFunnelChart", {
    enumerable: !0,
    get: function() {
        return funnel_1.registerFunnelChart;
    }
}), Object.defineProperty(exports, "FunnelChartSpecTransformer", {
    enumerable: !0,
    get: function() {
        return funnel_1.FunnelChartSpecTransformer;
    }
});

const gauge_1 = require("./gauge");

Object.defineProperty(exports, "GaugeChart", {
    enumerable: !0,
    get: function() {
        return gauge_1.GaugeChart;
    }
}), Object.defineProperty(exports, "registerGaugeChart", {
    enumerable: !0,
    get: function() {
        return gauge_1.registerGaugeChart;
    }
});

const heatmap_1 = require("./heatmap");

Object.defineProperty(exports, "HeatmapChart", {
    enumerable: !0,
    get: function() {
        return heatmap_1.HeatmapChart;
    }
}), Object.defineProperty(exports, "registerHeatmapChart", {
    enumerable: !0,
    get: function() {
        return heatmap_1.registerHeatmapChart;
    }
});

const histogram_1 = require("./histogram");

Object.defineProperty(exports, "HistogramChartSpecTransformer", {
    enumerable: !0,
    get: function() {
        return histogram_1.HistogramChartSpecTransformer;
    }
}), Object.defineProperty(exports, "HistogramChart", {
    enumerable: !0,
    get: function() {
        return histogram_1.HistogramChart;
    }
}), Object.defineProperty(exports, "registerHistogramChart", {
    enumerable: !0,
    get: function() {
        return histogram_1.registerHistogramChart;
    }
});

const line_1 = require("./line");

Object.defineProperty(exports, "LineChart", {
    enumerable: !0,
    get: function() {
        return line_1.LineChart;
    }
}), Object.defineProperty(exports, "registerLineChart", {
    enumerable: !0,
    get: function() {
        return line_1.registerLineChart;
    }
});

const map_1 = require("./map");

Object.defineProperty(exports, "MapChart", {
    enumerable: !0,
    get: function() {
        return map_1.MapChart;
    }
}), Object.defineProperty(exports, "registerMapChart", {
    enumerable: !0,
    get: function() {
        return map_1.registerMapChart;
    }
});

const pie_1 = require("./pie");

Object.defineProperty(exports, "BasePieChart", {
    enumerable: !0,
    get: function() {
        return pie_1.BasePieChart;
    }
}), Object.defineProperty(exports, "BasePieChartSpecTransformer", {
    enumerable: !0,
    get: function() {
        return pie_1.BasePieChartSpecTransformer;
    }
}), Object.defineProperty(exports, "PieChart", {
    enumerable: !0,
    get: function() {
        return pie_1.PieChart;
    }
}), Object.defineProperty(exports, "registerPieChart", {
    enumerable: !0,
    get: function() {
        return pie_1.registerPieChart;
    }
});

const circular_1 = require("./progress/circular");

Object.defineProperty(exports, "CircularProgressChart", {
    enumerable: !0,
    get: function() {
        return circular_1.CircularProgressChart;
    }
}), Object.defineProperty(exports, "registerCircularProgressChart", {
    enumerable: !0,
    get: function() {
        return circular_1.registerCircularProgressChart;
    }
});

const linear_1 = require("./progress/linear");

Object.defineProperty(exports, "LinearProgressChart", {
    enumerable: !0,
    get: function() {
        return linear_1.LinearProgressChart;
    }
}), Object.defineProperty(exports, "registerLinearProgressChart", {
    enumerable: !0,
    get: function() {
        return linear_1.registerLinearProgressChart;
    }
});

const radar_1 = require("./radar");

Object.defineProperty(exports, "RadarChart", {
    enumerable: !0,
    get: function() {
        return radar_1.RadarChart;
    }
}), Object.defineProperty(exports, "registerRadarChart", {
    enumerable: !0,
    get: function() {
        return radar_1.registerRadarChart;
    }
});

const range_area_1 = require("./range-area");

Object.defineProperty(exports, "RangeAreaChart", {
    enumerable: !0,
    get: function() {
        return range_area_1.RangeAreaChart;
    }
}), Object.defineProperty(exports, "registerRangeAreaChart", {
    enumerable: !0,
    get: function() {
        return range_area_1.registerRangeAreaChart;
    }
});

const range_column_1 = require("./range-column");

Object.defineProperty(exports, "RangeColumnChart", {
    enumerable: !0,
    get: function() {
        return range_column_1.RangeColumnChart;
    }
}), Object.defineProperty(exports, "registerRangeColumnChart", {
    enumerable: !0,
    get: function() {
        return range_column_1.registerRangeColumnChart;
    }
});

const rose_1 = require("./rose");

Object.defineProperty(exports, "RoseChart", {
    enumerable: !0,
    get: function() {
        return rose_1.RoseChart;
    }
}), Object.defineProperty(exports, "registerRoseChart", {
    enumerable: !0,
    get: function() {
        return rose_1.registerRoseChart;
    }
});

const sankey_1 = require("./sankey");

Object.defineProperty(exports, "SankeyChart", {
    enumerable: !0,
    get: function() {
        return sankey_1.SankeyChart;
    }
}), Object.defineProperty(exports, "registerSankeyChart", {
    enumerable: !0,
    get: function() {
        return sankey_1.registerSankeyChart;
    }
});

const scatter_1 = require("./scatter");

Object.defineProperty(exports, "ScatterChart", {
    enumerable: !0,
    get: function() {
        return scatter_1.ScatterChart;
    }
}), Object.defineProperty(exports, "registerScatterChart", {
    enumerable: !0,
    get: function() {
        return scatter_1.registerScatterChart;
    }
});

const sequence_1 = require("./sequence");

Object.defineProperty(exports, "SequenceChart", {
    enumerable: !0,
    get: function() {
        return sequence_1.SequenceChart;
    }
}), Object.defineProperty(exports, "registerSequenceChart", {
    enumerable: !0,
    get: function() {
        return sequence_1.registerSequenceChart;
    }
});

const sunburst_1 = require("./sunburst");

Object.defineProperty(exports, "SunburstChart", {
    enumerable: !0,
    get: function() {
        return sunburst_1.SunburstChart;
    }
}), Object.defineProperty(exports, "registerSunburstChart", {
    enumerable: !0,
    get: function() {
        return sunburst_1.registerSunburstChart;
    }
});

const treemap_1 = require("./treemap");

Object.defineProperty(exports, "TreemapChart", {
    enumerable: !0,
    get: function() {
        return treemap_1.TreemapChart;
    }
}), Object.defineProperty(exports, "registerTreemapChart", {
    enumerable: !0,
    get: function() {
        return treemap_1.registerTreemapChart;
    }
});

const waterfall_1 = require("./waterfall");

Object.defineProperty(exports, "WaterfallChart", {
    enumerable: !0,
    get: function() {
        return waterfall_1.WaterfallChart;
    }
}), Object.defineProperty(exports, "registerWaterfallChart", {
    enumerable: !0,
    get: function() {
        return waterfall_1.registerWaterfallChart;
    }
});

const correlation_1 = require("./correlation");

Object.defineProperty(exports, "CorrelationChart", {
    enumerable: !0,
    get: function() {
        return correlation_1.CorrelationChart;
    }
}), Object.defineProperty(exports, "registerCorrelationChart", {
    enumerable: !0,
    get: function() {
        return correlation_1.registerCorrelationChart;
    }
});

const liquid_1 = require("./liquid");

Object.defineProperty(exports, "LiquidChart", {
    enumerable: !0,
    get: function() {
        return liquid_1.LiquidChart;
    }
}), Object.defineProperty(exports, "registerLiquidChart", {
    enumerable: !0,
    get: function() {
        return liquid_1.registerLiquidChart;
    }
});

const word_cloud_1 = require("./word-cloud");

Object.defineProperty(exports, "WordCloudChart", {
    enumerable: !0,
    get: function() {
        return word_cloud_1.WordCloudChart;
    }
}), Object.defineProperty(exports, "registerWordCloudChart", {
    enumerable: !0,
    get: function() {
        return word_cloud_1.registerWordCloudChart;
    }
}), Object.defineProperty(exports, "registerWordCloudShapeChart", {
    enumerable: !0,
    get: function() {
        return word_cloud_1.registerWordCloudShapeChart;
    }
});

const base_chart_1 = require("./base/base-chart");

Object.defineProperty(exports, "BaseChart", {
    enumerable: !0,
    get: function() {
        return base_chart_1.BaseChart;
    }
});

const venn_1 = require("./venn");

Object.defineProperty(exports, "VennChart", {
    enumerable: !0,
    get: function() {
        return venn_1.VennChart;
    }
}), Object.defineProperty(exports, "registerVennChart", {
    enumerable: !0,
    get: function() {
        return venn_1.registerVennChart;
    }
});

const mosaic_1 = require("./mosaic");

Object.defineProperty(exports, "MosaicChart", {
    enumerable: !0,
    get: function() {
        return mosaic_1.MosaicChart;
    }
}), Object.defineProperty(exports, "registerMosaicChart", {
    enumerable: !0,
    get: function() {
        return mosaic_1.registerMosaicChart;
    }
});

const base_1 = require("./word-cloud/base/base");

Object.defineProperty(exports, "BaseWordCloudChart", {
    enumerable: !0,
    get: function() {
        return base_1.BaseWordCloudChart;
    }
});

const word_cloud_base_transformer_1 = require("./word-cloud/base/word-cloud-base-transformer");

Object.defineProperty(exports, "BaseWordCloudChartSpecTransformer", {
    enumerable: !0,
    get: function() {
        return word_cloud_base_transformer_1.BaseWordCloudChartSpecTransformer;
    }
});

const base_2 = require("./histogram/base/base");

Object.defineProperty(exports, "BaseHistogramChart", {
    enumerable: !0,
    get: function() {
        return base_2.BaseHistogramChart;
    }
});

var util_1 = require("./util");

Object.defineProperty(exports, "setDefaultCrosshairForCartesianChart", {
    enumerable: !0,
    get: function() {
        return util_1.setDefaultCrosshairForCartesianChart;
    }
});
