// @visactor/vchart-extension/cjs/charts/ranking-list/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.defaultSpec = void 0;

const cornerRadius = 5, animationDuration = 1e3, fontSize = 20, pageSize = 5, scrollSize = 1, primaryFontColor = "#21252c", shadowColor = "rgba(33,37,44,0.1)";

exports.defaultSpec = {
    labelLayout: "top",
    bar: {
        style: {
            cornerRadius: 5
        }
    },
    barBackground: {
        type: "rect",
        style: {
            fill: shadowColor,
            cornerRadius: 5
        }
    },
    rankingIcon: {
        visible: !0,
        style: {
            fill: "#21252c",
            size: 20
        }
    },
    nameLabel: {
        visible: !0,
        style: {
            fontSize: 20,
            fontWeight: "normal",
            fill: "#21252c",
            textBaseline: "middle"
        }
    },
    orderLabel: {
        visible: !0,
        style: {
            fontSize: 20,
            fontWeight: "normal",
            fill: "#21252c",
            textBaseline: "middle"
        }
    },
    valueLabel: {
        visible: !0,
        style: {
            fontSize: 20,
            fontWeight: "normal",
            fill: "#21252c",
            textBaseline: "middle"
        }
    },
    pageSize: 5,
    scrollSize: 1,
    animationAppear: {
        enable: !0,
        type: "grow",
        duration: 1e3,
        easing: "linear"
    },
    animationUpdate: {
        enable: !0,
        type: "grow",
        duration: 1e3,
        easing: "linear"
    },
    animationNormal: {
        interval: 1e3
    }
};
