// @visactor/vchart-extension/cjs/charts/ranking-list/ranking-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRankingList = exports.RankingList = void 0;

const vchart_1 = require("@visactor/vchart"), ranking_list_transformer_1 = require("./ranking-list-transformer");

class RankingList extends vchart_1.BaseChart {
    constructor() {
        super(...arguments), this.type = "rankingList", this.transformerConstructor = ranking_list_transformer_1.RankingListChartSpecTransformer;
    }
    init() {
        this.isValid() && super.init();
    }
    isValid() {
        var _a, _b, _c, _d;
        const {xField: xField, yField: yField, data: data} = this._spec;
        return xField && yField ? !!data || (null === (_d = (_c = this._option).onError) || void 0 === _d || _d.call(_c, "Data is required"), 
        !1) : (null === (_b = (_a = this._option).onError) || void 0 === _b || _b.call(_a, "Missing Required Config: `xField`, `yField` "), 
        !1);
    }
}

exports.RankingList = RankingList, RankingList.type = "rankingList", RankingList.view = "singleDefault", 
RankingList.transformerConstructor = ranking_list_transformer_1.RankingListChartSpecTransformer;

const registerRankingList = option => {
    const vchartConstructor = (null == option ? void 0 : option.VChart) || vchart_1.VChart;
    vchartConstructor && vchartConstructor.useChart([ RankingList, vchart_1.BarChart ]);
};

exports.registerRankingList = registerRankingList;
