// @visactor/vrender-core/es/common/event-listener-manager.js
export class EventListenerManager {
    constructor() {
        this._listenerMap = new Map, this._eventListenerTransformer = event => event;
    }
    setEventListenerTransformer(transformer) {
        this._eventListenerTransformer = transformer || (event => event);
    }
    addEventListener(type, listener, options) {
        if (!listener) return;
        const wrappedListener = event => {
            const transformedEvent = this._eventListenerTransformer(event);
            "function" == typeof listener ? listener(transformedEvent) : listener.handleEvent && listener.handleEvent(transformedEvent);
        };
        this._listenerMap.has(type) || this._listenerMap.set(type, new Map), this._listenerMap.get(type).set(listener, wrappedListener), 
        this._nativeAddEventListener(type, wrappedListener, options);
    }
    removeEventListener(type, listener, options) {
        var _a;
        if (!listener) return;
        const wrappedListener = null === (_a = this._listenerMap.get(type)) || void 0 === _a ? void 0 : _a.get(listener);
        wrappedListener && (this._nativeRemoveEventListener(type, wrappedListener, options), 
        this._listenerMap.get(type).delete(listener), 0 === this._listenerMap.get(type).size && this._listenerMap.delete(type));
    }
    dispatchEvent(event) {
        return this._nativeDispatchEvent(event);
    }
    clearAllEventListeners() {
        this._listenerMap.forEach(((listenersMap, type) => {
            listenersMap.forEach(((wrappedListener, originalListener) => {
                this._nativeRemoveEventListener(type, wrappedListener, void 0);
            }));
        })), this._listenerMap.clear();
    }
    _nativeAddEventListener(type, listener, options) {
        throw new Error("_nativeAddEventListener must be implemented by derived classes");
    }
    _nativeRemoveEventListener(type, listener, options) {
        throw new Error("_nativeRemoveEventListener must be implemented by derived classes");
    }
    _nativeDispatchEvent(event) {
        throw new Error("_nativeDispatchEvent must be implemented by derived classes");
    }
}
