// @visactor/vchart-extension/cjs/charts/word-cloud-3d/chart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerWordCloudShape3dChart = exports.registerWordCloud3dChart = exports.WordCloud3dChart = void 0;

const vchart_1 = require("@visactor/vchart"), plugin_1 = require("../3d/plugin"), series_1 = require("./series"), chart_spec_transformer_1 = require("./chart-spec-transformer"), enum_1 = require("../3d/enum");

class WordCloud3dChart extends vchart_1.BaseWordCloudChart {
    constructor() {
        super(...arguments), this.transformerConstructor = chart_spec_transformer_1.WordCloud3dChartSpecTransformer, 
        this.type = "wordCloud3d", this.seriesType = enum_1.SeriesType3dEnum.wordCloud3d;
    }
}

exports.WordCloud3dChart = WordCloud3dChart, WordCloud3dChart.type = "wordCloud3d", 
WordCloud3dChart.seriesType = enum_1.SeriesType3dEnum.wordCloud3d, WordCloud3dChart.transformerConstructor = chart_spec_transformer_1.WordCloud3dChartSpecTransformer;

const registerWordCloud3dChart = () => {
    (0, plugin_1.register3DPlugin)(), (0, series_1.registerWordCloud3dSeries)(), vchart_1.Factory.registerChart(WordCloud3dChart.type, WordCloud3dChart);
};

exports.registerWordCloud3dChart = registerWordCloud3dChart;

const registerWordCloudShape3dChart = () => {
    (0, plugin_1.register3DPlugin)(), (0, series_1.registerWordCloudShape3dSeries)(), 
    (0, exports.registerWordCloud3dChart)();
};

exports.registerWordCloudShape3dChart = registerWordCloudShape3dChart;
