// @visactor/vrender-components/cjs/marker/register.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.loadMarkPointComponent = exports.loadMarkArcAreaComponent = exports.loadMarkAreaComponent = exports.loadMarkArcLineComponent = exports.loadMarkLineComponent = void 0;

const vrender_kits_1 = require("@visactor/vrender-kits"), register_1 = require("../tag/register"), register_2 = require("../segment/register"), vrender_animate_1 = require("@visactor/vrender-animate");

function loadBaseMarker() {
    (0, vrender_kits_1.registerGroup)(), (0, register_1.loadTagComponent)(), (0, vrender_animate_1.registerAnimate)();
}

function loadMarkLineComponent() {
    loadBaseMarker(), (0, register_2.loadSegmentComponent)(), (0, vrender_animate_1.registerAnimate)();
}

function loadMarkArcLineComponent() {
    loadBaseMarker(), (0, register_2.loadArcSegmentComponent)(), (0, vrender_animate_1.registerAnimate)();
}

function loadMarkAreaComponent() {
    loadBaseMarker(), (0, vrender_kits_1.registerPolygon)(), (0, vrender_animate_1.registerAnimate)();
}

function loadMarkArcAreaComponent() {
    loadBaseMarker(), (0, vrender_kits_1.registerArc)(), (0, vrender_animate_1.registerAnimate)();
}

function loadMarkPointComponent() {
    loadBaseMarker(), (0, register_2.loadSegmentComponent)(), (0, register_2.loadArcSegmentComponent)(), 
    (0, vrender_kits_1.registerSymbol)(), (0, vrender_kits_1.registerImage)(), (0, vrender_kits_1.registerLine)(), 
    (0, vrender_animate_1.registerAnimate)();
}

exports.loadMarkLineComponent = loadMarkLineComponent, exports.loadMarkArcLineComponent = loadMarkArcLineComponent, 
exports.loadMarkAreaComponent = loadMarkAreaComponent, exports.loadMarkArcAreaComponent = loadMarkArcAreaComponent, 
exports.loadMarkPointComponent = loadMarkPointComponent;
