// @visactor/react-vchart/cjs/register.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerHistogramChart = exports.registerHeatmapChart = exports.registerGaugeChart = exports.registerFunnelChart = exports.registerCommonChart = exports.registerCircularProgressChart = exports.registerCirclePackingChart = exports.registerBoxplotChart = exports.registerBarChart = exports.registerAreaChart = exports.registerLineChart = exports.registerPictogramChart = exports.registerWordCloud3dChart = exports.registerRangeColumn3dChart = exports.registerPie3dChart = exports.registerHistogram3dChart = exports.registerFunnel3dChart = exports.registerBar3dChart = exports.registerAxis3dPlugin = exports.registerMapLabel = exports.registerPoptip = exports.registerPolarLinearAxis = exports.registerPolarCrossHair = exports.registerPolarBandAxis = exports.registerPlayer = exports.registerGeoMarkPoint = exports.registerPolarMarkPoint = exports.registerPolarMarkArea = exports.registerPolarMarkLine = exports.registerMarkPoint = exports.registerMarkArea = exports.registerMarkLine = exports.registerTotalLabel = exports.registerLabel = exports.registerIndicator = exports.registerGeoCoordinate = exports.registerDiscreteLegend = exports.registerDataZoom = exports.registerCustomMark = exports.registerContinuousLegend = exports.registerCartesianTimeAxis = exports.registerCartesianSymlogAxis = exports.registerCartesianLogAxis = exports.registerCartesianLinearAxis = exports.registerCartesianCrossHair = exports.registerCartesianBandAxis = exports.registerTooltip = exports.registerTitle = exports.registerScrollBar = exports.registerBrush = void 0, 
exports.registerMosaicChart = exports.registerVennChart = exports.registerWordCloudShapeChart = exports.registerLiquidChart = exports.registerCorrelationChart = exports.registerWordCloudChart = exports.registerWaterfallChart = exports.registerTreemapChart = exports.registerSunburstChart = exports.registerSequenceChart = exports.registerScatterChart = exports.registerSankeyChart = exports.registerRoseChart = exports.registerRangeColumnChart = exports.registerRangeAreaChart = exports.registerRadarChart = exports.registerPieChart = exports.registerMapChart = exports.registerLinearProgressChart = void 0;

var vchart_1 = require("@visactor/vchart");

Object.defineProperty(exports, "registerBrush", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerBrush;
    }
}), Object.defineProperty(exports, "registerScrollBar", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerScrollBar;
    }
}), Object.defineProperty(exports, "registerTitle", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerTitle;
    }
}), Object.defineProperty(exports, "registerTooltip", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerTooltip;
    }
}), Object.defineProperty(exports, "registerCartesianBandAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianBandAxis;
    }
}), Object.defineProperty(exports, "registerCartesianCrossHair", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianCrossHair;
    }
}), Object.defineProperty(exports, "registerCartesianLinearAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianLinearAxis;
    }
}), Object.defineProperty(exports, "registerCartesianLogAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianLogAxis;
    }
}), Object.defineProperty(exports, "registerCartesianSymlogAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianSymlogAxis;
    }
}), Object.defineProperty(exports, "registerCartesianTimeAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCartesianTimeAxis;
    }
}), Object.defineProperty(exports, "registerContinuousLegend", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerContinuousLegend;
    }
}), Object.defineProperty(exports, "registerCustomMark", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerCustomMark;
    }
}), Object.defineProperty(exports, "registerDataZoom", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerDataZoom;
    }
}), Object.defineProperty(exports, "registerDiscreteLegend", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerDiscreteLegend;
    }
}), Object.defineProperty(exports, "registerGeoCoordinate", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerGeoCoordinate;
    }
}), Object.defineProperty(exports, "registerIndicator", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerIndicator;
    }
}), Object.defineProperty(exports, "registerLabel", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerLabel;
    }
}), Object.defineProperty(exports, "registerTotalLabel", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerTotalLabel;
    }
}), Object.defineProperty(exports, "registerMarkLine", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerMarkLine;
    }
}), Object.defineProperty(exports, "registerMarkArea", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerMarkArea;
    }
}), Object.defineProperty(exports, "registerMarkPoint", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerMarkPoint;
    }
}), Object.defineProperty(exports, "registerPolarMarkLine", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarMarkLine;
    }
}), Object.defineProperty(exports, "registerPolarMarkArea", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarMarkArea;
    }
}), Object.defineProperty(exports, "registerPolarMarkPoint", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarMarkPoint;
    }
}), Object.defineProperty(exports, "registerGeoMarkPoint", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerGeoMarkPoint;
    }
}), Object.defineProperty(exports, "registerPlayer", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPlayer;
    }
}), Object.defineProperty(exports, "registerPolarBandAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarBandAxis;
    }
}), Object.defineProperty(exports, "registerPolarCrossHair", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarCrossHair;
    }
}), Object.defineProperty(exports, "registerPolarLinearAxis", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPolarLinearAxis;
    }
}), Object.defineProperty(exports, "registerPoptip", {
    enumerable: !0,
    get: function() {
        return vchart_1.registerPoptip;
    }
});

var vchart_extension_1 = require("@visactor/vchart-extension");

Object.defineProperty(exports, "registerMapLabel", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerMapLabel;
    }
}), Object.defineProperty(exports, "registerAxis3dPlugin", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerAxis3dPlugin;
    }
}), Object.defineProperty(exports, "registerBar3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerBar3dChart;
    }
}), Object.defineProperty(exports, "registerFunnel3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerFunnel3dChart;
    }
}), Object.defineProperty(exports, "registerHistogram3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerHistogram3dChart;
    }
}), Object.defineProperty(exports, "registerPie3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerPie3dChart;
    }
}), Object.defineProperty(exports, "registerRangeColumn3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerRangeColumn3dChart;
    }
}), Object.defineProperty(exports, "registerWordCloud3dChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerWordCloud3dChart;
    }
}), Object.defineProperty(exports, "registerPictogramChart", {
    enumerable: !0,
    get: function() {
        return vchart_extension_1.registerPictogramChart;
    }
});

var vchart_2 = require("@visactor/vchart");

Object.defineProperty(exports, "registerLineChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerLineChart;
    }
}), Object.defineProperty(exports, "registerAreaChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerAreaChart;
    }
}), Object.defineProperty(exports, "registerBarChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerBarChart;
    }
}), Object.defineProperty(exports, "registerBoxplotChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerBoxplotChart;
    }
}), Object.defineProperty(exports, "registerCirclePackingChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerCirclePackingChart;
    }
}), Object.defineProperty(exports, "registerCircularProgressChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerCircularProgressChart;
    }
}), Object.defineProperty(exports, "registerCommonChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerCommonChart;
    }
}), Object.defineProperty(exports, "registerFunnelChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerFunnelChart;
    }
}), Object.defineProperty(exports, "registerGaugeChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerGaugeChart;
    }
}), Object.defineProperty(exports, "registerHeatmapChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerHeatmapChart;
    }
}), Object.defineProperty(exports, "registerHistogramChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerHistogramChart;
    }
}), Object.defineProperty(exports, "registerLinearProgressChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerLinearProgressChart;
    }
}), Object.defineProperty(exports, "registerMapChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerMapChart;
    }
}), Object.defineProperty(exports, "registerPieChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerPieChart;
    }
}), Object.defineProperty(exports, "registerRadarChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerRadarChart;
    }
}), Object.defineProperty(exports, "registerRangeAreaChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerRangeAreaChart;
    }
}), Object.defineProperty(exports, "registerRangeColumnChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerRangeColumnChart;
    }
}), Object.defineProperty(exports, "registerRoseChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerRoseChart;
    }
}), Object.defineProperty(exports, "registerSankeyChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerSankeyChart;
    }
}), Object.defineProperty(exports, "registerScatterChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerScatterChart;
    }
}), Object.defineProperty(exports, "registerSequenceChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerSequenceChart;
    }
}), Object.defineProperty(exports, "registerSunburstChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerSunburstChart;
    }
}), Object.defineProperty(exports, "registerTreemapChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerTreemapChart;
    }
}), Object.defineProperty(exports, "registerWaterfallChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerWaterfallChart;
    }
}), Object.defineProperty(exports, "registerWordCloudChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerWordCloudChart;
    }
}), Object.defineProperty(exports, "registerCorrelationChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerCorrelationChart;
    }
}), Object.defineProperty(exports, "registerLiquidChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerLiquidChart;
    }
}), Object.defineProperty(exports, "registerWordCloudShapeChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerWordCloudShapeChart;
    }
}), Object.defineProperty(exports, "registerVennChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerVennChart;
    }
}), Object.defineProperty(exports, "registerMosaicChart", {
    enumerable: !0,
    get: function() {
        return vchart_2.registerMosaicChart;
    }
});
