// @visactor/vchart/cjs/mark/cell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerCellMark = exports.CellMark = void 0;

const factory_1 = require("./../core/factory"), vrender_kits_1 = require("@visactor/vrender-kits"), vrender_components_1 = require("@visactor/vrender-components"), vrender_core_1 = require("@visactor/vrender-core"), vutils_1 = require("@visactor/vutils"), base_1 = require("./base");

class CellMark extends base_1.BaseMark {
    constructor() {
        super(...arguments), this.type = CellMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            fill: void 0,
            lineWidth: 0,
            padding: 0
        });
    }
    _createGraphic(attrs = {}) {
        return (0, vrender_core_1.createSymbol)(attrs);
    }
    _transformGraphicAttributes(g, attrs, groupAttrs) {
        var _a, _b, _c;
        const symbolAttrs = super._transformGraphicAttributes(g, attrs, groupAttrs), symbolType = null !== (_b = null !== (_a = symbolAttrs.shape) && void 0 !== _a ? _a : symbolAttrs.symbolType) && void 0 !== _b ? _b : null === (_c = g.attribute) || void 0 === _c ? void 0 : _c.symbolType;
        return (0, vutils_1.isNil)(symbolType) ? symbolAttrs.symbolType = "rect" : symbolAttrs.symbolType = symbolType, 
        symbolAttrs;
    }
}

exports.CellMark = CellMark, CellMark.type = "cell";

const registerCellMark = () => {
    factory_1.Factory.registerMark(CellMark.type, CellMark), factory_1.Factory.registerGraphicComponent("symbol", vrender_core_1.createSymbol), 
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerSymbol)(), 
    (0, vrender_components_1.registerSymbolDataLabel)();
};

exports.registerCellMark = registerCellMark;
