// @visactor/vchart/cjs/mark/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerAllMarks = exports.registerMarkFilterTransform = exports.registerMarkMapTransform = exports.registerComponentMark = exports.registerImageMark = exports.registerRippleMark = exports.registerGlyphMark = exports.registerPolygonMark = exports.registerArcMark = exports.registerPathMark = exports.registerRectMark = exports.registerAreaMark = exports.registerTextMark = exports.registerRuleMark = exports.registerGroupMark = exports.registerSymbolMark = exports.registerLineMark = exports.BaseMark = exports.ImageMark = exports.RippleMark = exports.PolygonMark = exports.BasePolygonMark = exports.ComponentMark = exports.ArcMark = exports.BaseArcMark = exports.GlyphMark = exports.PathMark = exports.RectMark = exports.AreaMark = exports.TextMark = exports.RuleMark = exports.GroupMark = exports.SymbolMark = exports.LineMark = exports.MarkTypeEnum = void 0;

const line_1 = require("./line");

Object.defineProperty(exports, "LineMark", {
    enumerable: !0,
    get: function() {
        return line_1.LineMark;
    }
}), Object.defineProperty(exports, "registerLineMark", {
    enumerable: !0,
    get: function() {
        return line_1.registerLineMark;
    }
});

const symbol_1 = require("./symbol");

Object.defineProperty(exports, "SymbolMark", {
    enumerable: !0,
    get: function() {
        return symbol_1.SymbolMark;
    }
}), Object.defineProperty(exports, "registerSymbolMark", {
    enumerable: !0,
    get: function() {
        return symbol_1.registerSymbolMark;
    }
});

const group_1 = require("./group");

Object.defineProperty(exports, "GroupMark", {
    enumerable: !0,
    get: function() {
        return group_1.GroupMark;
    }
}), Object.defineProperty(exports, "registerGroupMark", {
    enumerable: !0,
    get: function() {
        return group_1.registerGroupMark;
    }
});

const rule_1 = require("./rule");

Object.defineProperty(exports, "RuleMark", {
    enumerable: !0,
    get: function() {
        return rule_1.RuleMark;
    }
}), Object.defineProperty(exports, "registerRuleMark", {
    enumerable: !0,
    get: function() {
        return rule_1.registerRuleMark;
    }
});

const text_1 = require("./text");

Object.defineProperty(exports, "TextMark", {
    enumerable: !0,
    get: function() {
        return text_1.TextMark;
    }
}), Object.defineProperty(exports, "registerTextMark", {
    enumerable: !0,
    get: function() {
        return text_1.registerTextMark;
    }
});

const area_1 = require("./area");

Object.defineProperty(exports, "AreaMark", {
    enumerable: !0,
    get: function() {
        return area_1.AreaMark;
    }
}), Object.defineProperty(exports, "registerAreaMark", {
    enumerable: !0,
    get: function() {
        return area_1.registerAreaMark;
    }
});

const rect_1 = require("./rect");

Object.defineProperty(exports, "RectMark", {
    enumerable: !0,
    get: function() {
        return rect_1.RectMark;
    }
}), Object.defineProperty(exports, "registerRectMark", {
    enumerable: !0,
    get: function() {
        return rect_1.registerRectMark;
    }
});

const path_1 = require("./path");

Object.defineProperty(exports, "PathMark", {
    enumerable: !0,
    get: function() {
        return path_1.PathMark;
    }
}), Object.defineProperty(exports, "registerPathMark", {
    enumerable: !0,
    get: function() {
        return path_1.registerPathMark;
    }
});

const arc_1 = require("./arc");

Object.defineProperty(exports, "ArcMark", {
    enumerable: !0,
    get: function() {
        return arc_1.ArcMark;
    }
}), Object.defineProperty(exports, "BaseArcMark", {
    enumerable: !0,
    get: function() {
        return arc_1.BaseArcMark;
    }
}), Object.defineProperty(exports, "registerArcMark", {
    enumerable: !0,
    get: function() {
        return arc_1.registerArcMark;
    }
});

const component_1 = require("./component");

Object.defineProperty(exports, "ComponentMark", {
    enumerable: !0,
    get: function() {
        return component_1.ComponentMark;
    }
}), Object.defineProperty(exports, "registerComponentMark", {
    enumerable: !0,
    get: function() {
        return component_1.registerComponentMark;
    }
});

const ripple_1 = require("./ripple");

Object.defineProperty(exports, "RippleMark", {
    enumerable: !0,
    get: function() {
        return ripple_1.RippleMark;
    }
}), Object.defineProperty(exports, "registerRippleMark", {
    enumerable: !0,
    get: function() {
        return ripple_1.registerRippleMark;
    }
});

const glyph_1 = require("./glyph");

Object.defineProperty(exports, "GlyphMark", {
    enumerable: !0,
    get: function() {
        return glyph_1.GlyphMark;
    }
}), Object.defineProperty(exports, "registerGlyphMark", {
    enumerable: !0,
    get: function() {
        return glyph_1.registerGlyphMark;
    }
});

const base_1 = require("./base");

Object.defineProperty(exports, "BaseMark", {
    enumerable: !0,
    get: function() {
        return base_1.BaseMark;
    }
});

const polygon_1 = require("./polygon/polygon");

Object.defineProperty(exports, "PolygonMark", {
    enumerable: !0,
    get: function() {
        return polygon_1.PolygonMark;
    }
}), Object.defineProperty(exports, "registerPolygonMark", {
    enumerable: !0,
    get: function() {
        return polygon_1.registerPolygonMark;
    }
});

const image_1 = require("./image");

Object.defineProperty(exports, "ImageMark", {
    enumerable: !0,
    get: function() {
        return image_1.ImageMark;
    }
}), Object.defineProperty(exports, "registerImageMark", {
    enumerable: !0,
    get: function() {
        return image_1.registerImageMark;
    }
});

const filter_1 = require("./transform/filter");

Object.defineProperty(exports, "registerMarkFilterTransform", {
    enumerable: !0,
    get: function() {
        return filter_1.registerMarkFilterTransform;
    }
});

const map_1 = require("./transform/map");

Object.defineProperty(exports, "registerMarkMapTransform", {
    enumerable: !0,
    get: function() {
        return map_1.registerMarkMapTransform;
    }
});

const base_polygon_1 = require("./polygon/base-polygon");

Object.defineProperty(exports, "BasePolygonMark", {
    enumerable: !0,
    get: function() {
        return base_polygon_1.BasePolygonMark;
    }
});

const type_1 = require("./interface/type");

Object.defineProperty(exports, "MarkTypeEnum", {
    enumerable: !0,
    get: function() {
        return type_1.MarkTypeEnum;
    }
});

const registerAllMarks = () => {
    (0, group_1.registerGroupMark)(), (0, line_1.registerLineMark)(), (0, symbol_1.registerSymbolMark)(), 
    (0, rule_1.registerRuleMark)(), (0, text_1.registerTextMark)(), (0, area_1.registerAreaMark)(), 
    (0, rect_1.registerRectMark)(), (0, path_1.registerPathMark)(), (0, arc_1.registerArcMark)(), 
    (0, polygon_1.registerPolygonMark)(), (0, image_1.registerImageMark)(), (0, ripple_1.registerRippleMark)(), 
    (0, component_1.registerComponentMark)();
};

exports.registerAllMarks = registerAllMarks;
