// @visactor/vchart/cjs/mark/label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerLabelMark = exports.LabelMark = void 0;

const factory_1 = require("./../core/factory"), text_1 = require("./text");

class LabelMark extends text_1.TextMark {
    constructor() {
        super(...arguments), this.skipEncode = !1;
    }
    getRule() {
        return this._rule;
    }
    setRule(rule) {
        this._rule = rule;
    }
    getTarget() {
        return this._target;
    }
    setTarget(target) {
        this._target = target, this._rule || this.setRule(target.type);
    }
    getComponent() {
        return this._component;
    }
    setComponent(component) {
        this._component = component;
    }
    render() {}
}

exports.LabelMark = LabelMark, LabelMark.type = "text", LabelMark.constructorType = "label";

const registerLabelMark = () => {
    (0, text_1.registerTextMark)(), factory_1.Factory.registerMark(LabelMark.constructorType, LabelMark);
};

exports.registerLabelMark = registerLabelMark;
