// @visactor/vchart/cjs/mark/rect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRectMark = exports.RectMark = void 0;

const factory_1 = require("./../core/factory"), base_mark_1 = require("./base/base-mark"), config_1 = require("../animation/config"), vrender_kits_1 = require("@visactor/vrender-kits"), vrender_components_1 = require("@visactor/vrender-components"), vrender_core_1 = require("@visactor/vrender-core");

class RectMark extends base_mark_1.BaseMark {
    constructor() {
        super(...arguments), this.type = RectMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            width: void 0,
            height: void 0,
            lineWidth: 0
        });
    }
}

exports.RectMark = RectMark, RectMark.type = "rect";

const registerRectMark = () => {
    (0, vrender_kits_1.registerShadowRoot)(), (0, vrender_kits_1.registerRect)(), (0, 
    vrender_components_1.registerRectDataLabel)(), factory_1.Factory.registerMark(RectMark.type, RectMark), 
    (0, config_1.registerRectAnimation)(), factory_1.Factory.registerGraphicComponent("rect", vrender_core_1.createRect);
};

exports.registerRectMark = registerRectMark;
