// @visactor/vchart/cjs/mark/rule.js
"use strict";

var __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRuleMark = exports.RuleMark = void 0;

const factory_1 = require("./../core/factory"), base_mark_1 = require("./base/base-mark"), vrender_kits_1 = require("@visactor/vrender-kits"), vrender_core_1 = require("@visactor/vrender-core");

class RuleMark extends base_mark_1.BaseMark {
    constructor() {
        super(...arguments), this.type = RuleMark.type;
    }
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            x1: 0,
            y1: 0
        });
    }
    _transformGraphicAttributes(g, attrs, groupAttrs) {
        const finalAttrs = super._transformGraphicAttributes(g, attrs, groupAttrs), {x: x, x1: x1, y: y, y1: y1} = finalAttrs, rest = __rest(finalAttrs, [ "x", "x1", "y", "y1" ]);
        return Object.assign(Object.assign({}, rest), {
            points: [ {
                x: x,
                y: y
            }, {
                x: x1,
                y: y1
            } ]
        });
    }
    _createGraphic(attrs = {}) {
        return (0, vrender_core_1.createLine)(attrs);
    }
}

exports.RuleMark = RuleMark, RuleMark.type = "rule";

const registerRuleMark = () => {
    factory_1.Factory.registerMark(RuleMark.type, RuleMark), (0, vrender_kits_1.registerShadowRoot)(), 
    (0, vrender_kits_1.registerLine)(), factory_1.Factory.registerGraphicComponent("line", vrender_core_1.createLine);
};

exports.registerRuleMark = registerRuleMark;
