// @visactor/vchart/esm/constant/event.js
export var HOOK_EVENT;

!function(HOOK_EVENT) {
    HOOK_EVENT.BEFORE_EVALUATE_DATA = "beforeEvaluateData", HOOK_EVENT.AFTER_EVALUATE_DATA = "afterEvaluateData", 
    HOOK_EVENT.BEFORE_EVALUATE_SCALE = "beforeEvaluateScale", HOOK_EVENT.AFTER_EVALUATE_SCALE = "afterEvaluateScale", 
    HOOK_EVENT.BEFORE_PARSE_VIEW = "beforeParseView", HOOK_EVENT.AFTER_PARSE_VIEW = "afterParseView", 
    HOOK_EVENT.BEFORE_TRANSFORM = "beforeTransform", HOOK_EVENT.AFTER_TRANSFORM = "afterTransform", 
    HOOK_EVENT.BEFORE_CREATE_VRENDER_STAGE = "beforeCreateVRenderStage", HOOK_EVENT.AFTER_CREATE_VRENDER_STAGE = "afterCreateVRenderStage", 
    HOOK_EVENT.BEFORE_CREATE_VRENDER_LAYER = "beforeCreateVRenderLayer", HOOK_EVENT.AFTER_CREATE_VRENDER_LAYER = "afterCreateVRenderLayer", 
    HOOK_EVENT.BEFORE_STAGE_RESIZE = "beforeStageResize", HOOK_EVENT.AFTER_STAGE_RESIZE = "afterStageResize", 
    HOOK_EVENT.BEFORE_VRENDER_DRAW = "beforeVRenderDraw", HOOK_EVENT.AFTER_VRENDER_DRAW = "afterVRenderDraw", 
    HOOK_EVENT.BEFORE_MARK_JOIN = "beforeMarkJoin", HOOK_EVENT.AFTER_MARK_JOIN = "afterMarkJoin", 
    HOOK_EVENT.BEFORE_MARK_UPDATE = "beforeMarkUpdate", HOOK_EVENT.AFTER_MARK_UPDATE = "afterMarkUpdate", 
    HOOK_EVENT.BEFORE_MARK_STATE = "beforeMarkState", HOOK_EVENT.AFTER_MARK_STATE = "afterMarkState", 
    HOOK_EVENT.BEFORE_MARK_ENCODE = "beforeMarkEncode", HOOK_EVENT.AFTER_MARK_ENCODE = "afterMarkEncode", 
    HOOK_EVENT.BEFORE_DO_LAYOUT = "beforeDoLayout", HOOK_EVENT.AFTER_DO_LAYOUT = "afterDoLayout", 
    HOOK_EVENT.BEFORE_MARK_LAYOUT_END = "beforeMarkLayoutEnd", HOOK_EVENT.AFTER_MARK_LAYOUT_END = "afterMarkLayoutEnd", 
    HOOK_EVENT.BEFORE_DO_RENDER = "beforeDoRender", HOOK_EVENT.AFTER_DO_RENDER = "afterDoRender", 
    HOOK_EVENT.BEFORE_MARK_RENDER_END = "beforeMarkRenderEnd", HOOK_EVENT.AFTER_MARK_RENDER_END = "afterMarkRenderEnd", 
    HOOK_EVENT.BEFORE_CREATE_VRENDER_MARK = "beforeCreateVRenderMark", HOOK_EVENT.AFTER_CREATE_VRENDER_MARK = "afterCreateVRenderMark", 
    HOOK_EVENT.BEFORE_ADD_VRENDER_MARK = "beforeAddVRenderMark", HOOK_EVENT.AFTER_ADD_VRENDER_MARK = "afterAddVRenderMark", 
    HOOK_EVENT.AFTER_VRENDER_NEXT_RENDER = "afterVRenderNextRender", HOOK_EVENT.BEFORE_ELEMENT_UPDATE_DATA = "beforeElementUpdateData", 
    HOOK_EVENT.AFTER_ELEMENT_UPDATE_DATA = "afterElementUpdateData", HOOK_EVENT.BEFORE_ELEMENT_STATE = "beforeElementState", 
    HOOK_EVENT.AFTER_ELEMENT_STATE = "afterElementState", HOOK_EVENT.BEFORE_ELEMENT_ENCODE = "beforeElementEncode", 
    HOOK_EVENT.AFTER_ELEMENT_ENCODE = "afterElementEncode", HOOK_EVENT.ANIMATION_START = "animationStart", 
    HOOK_EVENT.ANIMATION_END = "animationEnd", HOOK_EVENT.ELEMENT_ANIMATION_START = "elementAnimationStart", 
    HOOK_EVENT.ELEMENT_ANIMATION_END = "elementAnimationEnd", HOOK_EVENT.ALL_ANIMATION_START = "allAnimationStart", 
    HOOK_EVENT.ALL_ANIMATION_END = "allAnimationEnd";
}(HOOK_EVENT || (HOOK_EVENT = {}));

export const BASE_EVENTS = [ "pointerdown", "pointerup", "pointerupoutside", "pointertap", "pointerover", "pointermove", "pointerenter", "pointerleave", "pointerout", "mousedown", "mouseup", "mouseupoutside", "rightdown", "rightup", "rightupoutside", "click", "dblclick", "mousemove", "mouseover", "mouseout", "mouseenter", "mouseleave", "wheel", "touchstart", "touchend", "touchendoutside", "touchmove", "touchcancel", "tap", "dragstart", "drag", "dragenter", "dragleave", "dragover", "dragend", "drop", "pan", "panstart", "panend", "press", "pressup", "pressend", "pinch", "pinchstart", "pinchend", "swipe" ];

export var ChartEvent;

!function(ChartEvent) {
    ChartEvent.initialized = "initialized", ChartEvent.rendered = "rendered", ChartEvent.renderFinished = "renderFinished", 
    ChartEvent.animationFinished = "animationFinished", ChartEvent.regionSeriesDataFilterOver = "regionSeriesDataFilterOver", 
    ChartEvent.afterInitData = "afterInitData", ChartEvent.afterInitEvent = "afterInitEvent", 
    ChartEvent.afterInitMark = "afterInitMark", ChartEvent.rawDataUpdate = "rawDataUpdate", 
    ChartEvent.viewDataFilterOver = "viewDataFilterOver", ChartEvent.viewDataUpdate = "viewDataUpdate", 
    ChartEvent.viewDataStatisticsUpdate = "viewDataStatisticsUpdate", ChartEvent.markDeltaYUpdate = "markDeltaYUpdate", 
    ChartEvent.viewDataLabelUpdate = "viewDataLabelUpdate", ChartEvent.scaleDomainUpdate = "scaleDomainUpdate", 
    ChartEvent.scaleUpdate = "scaleUpdate", ChartEvent.dataZoomChange = "dataZoomChange", 
    ChartEvent.drill = "drill", ChartEvent.layoutStart = "layoutStart", ChartEvent.layoutEnd = "layoutEnd", 
    ChartEvent.layoutRectUpdate = "layoutRectUpdate", ChartEvent.playerPlay = "playerPlay", 
    ChartEvent.playerPause = "playerPause", ChartEvent.playerEnd = "playerEnd", ChartEvent.playerChange = "playerChange", 
    ChartEvent.playerForward = "playerForward", ChartEvent.playerBackward = "playerBackward", 
    ChartEvent.scrollBarChange = "scrollBarChange", ChartEvent.brushStart = "brushStart", 
    ChartEvent.brushChange = "brushChange", ChartEvent.brushEnd = "brushEnd", ChartEvent.brushClear = "brushClear", 
    ChartEvent.brushActive = "brushActive", ChartEvent.legendSelectedDataChange = "legendSelectedDataChange", 
    ChartEvent.legendFilter = "legendFilter", ChartEvent.legendItemClick = "legendItemClick", 
    ChartEvent.legendItemHover = "legendItemHover", ChartEvent.legendItemUnHover = "legendItemUnHover", 
    ChartEvent.tooltipShow = "tooltipShow", ChartEvent.tooltipHide = "tooltipHide", 
    ChartEvent.tooltipRelease = "tooltipRelease", ChartEvent.afterResize = "afterResize", 
    ChartEvent.afterRender = "afterRender", ChartEvent.afterLayout = "afterLayout", 
    ChartEvent.afterMarkLayoutEnd = "afterMarkLayoutEnd";
}(ChartEvent || (ChartEvent = {}));

export var Event_Source_Type;

!function(Event_Source_Type) {
    Event_Source_Type.chart = "chart", Event_Source_Type.window = "window", Event_Source_Type.canvas = "canvas";
}(Event_Source_Type || (Event_Source_Type = {}));

export var Event_Bubble_Level;

!function(Event_Bubble_Level) {
    Event_Bubble_Level.vchart = "vchart", Event_Bubble_Level.chart = "chart", Event_Bubble_Level.model = "model", 
    Event_Bubble_Level.mark = "mark";
}(Event_Bubble_Level || (Event_Bubble_Level = {}));

export const EventBubbleLevels = [ Event_Bubble_Level.mark, Event_Bubble_Level.model, Event_Bubble_Level.chart, Event_Bubble_Level.vchart ];
