// @visactor/vchart/cjs/interaction/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerElementHighlightByName = exports.registerElementHighlightByLegend = exports.registerElementHighlightByKey = exports.registerElementHighlightByGroup = exports.registerElementActiveByLegend = exports.registerElementActive = exports.registerDimensionHover = exports.ElementSelect = exports.registerElementSelect = exports.ElementHighlight = exports.registerElementHighlight = void 0;

var element_highlight_1 = require("./triggers/element-highlight");

Object.defineProperty(exports, "registerElementHighlight", {
    enumerable: !0,
    get: function() {
        return element_highlight_1.registerElementHighlight;
    }
}), Object.defineProperty(exports, "ElementHighlight", {
    enumerable: !0,
    get: function() {
        return element_highlight_1.ElementHighlight;
    }
});

var element_select_1 = require("./triggers/element-select");

Object.defineProperty(exports, "registerElementSelect", {
    enumerable: !0,
    get: function() {
        return element_select_1.registerElementSelect;
    }
}), Object.defineProperty(exports, "ElementSelect", {
    enumerable: !0,
    get: function() {
        return element_select_1.ElementSelect;
    }
});

var dimension_hover_1 = require("./triggers/dimension-hover");

Object.defineProperty(exports, "registerDimensionHover", {
    enumerable: !0,
    get: function() {
        return dimension_hover_1.registerDimensionHover;
    }
});

var element_active_1 = require("./triggers/element-active");

Object.defineProperty(exports, "registerElementActive", {
    enumerable: !0,
    get: function() {
        return element_active_1.registerElementActive;
    }
});

var element_active_by_legend_1 = require("./triggers/element-active-by-legend");

Object.defineProperty(exports, "registerElementActiveByLegend", {
    enumerable: !0,
    get: function() {
        return element_active_by_legend_1.registerElementActiveByLegend;
    }
});

var element_highlight_by_group_1 = require("./triggers/element-highlight-by-group");

Object.defineProperty(exports, "registerElementHighlightByGroup", {
    enumerable: !0,
    get: function() {
        return element_highlight_by_group_1.registerElementHighlightByGroup;
    }
});

var element_highlight_by_key_1 = require("./triggers/element-highlight-by-key");

Object.defineProperty(exports, "registerElementHighlightByKey", {
    enumerable: !0,
    get: function() {
        return element_highlight_by_key_1.registerElementHighlightByKey;
    }
});

var element_highlight_by_legend_1 = require("./triggers/element-highlight-by-legend");

Object.defineProperty(exports, "registerElementHighlightByLegend", {
    enumerable: !0,
    get: function() {
        return element_highlight_by_legend_1.registerElementHighlightByLegend;
    }
});

var element_highlight_by_name_1 = require("./triggers/element-highlight-by-name");

Object.defineProperty(exports, "registerElementHighlightByName", {
    enumerable: !0,
    get: function() {
        return element_highlight_by_name_1.registerElementHighlightByName;
    }
});
