// @visactor/vchart/cjs/theme/builtin/common/series/funnel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.funnel = exports.getFunnelTheme = void 0;

const getFunnelTheme = is3d => {
    const res = {
        label: {
            style: {
                fill: "white",
                textBaseline: "middle",
                lineWidth: 2
            }
        },
        outerLabel: {
            style: {
                fontSize: {
                    type: "token",
                    key: "l4FontSize"
                },
                fill: {
                    type: "palette",
                    key: "secondaryFontColor"
                }
            },
            line: {
                style: {
                    stroke: {
                        type: "palette",
                        key: "axisDomainColor"
                    }
                }
            }
        },
        transformLabel: {
            style: {
                fontSize: {
                    type: "token",
                    key: "l4FontSize"
                },
                fill: {
                    type: "palette",
                    key: "secondaryFontColor"
                },
                textBaseline: "middle"
            }
        }
    };
    return res[is3d ? "transform3d" : "transform"] = {
        style: {
            fill: {
                type: "palette",
                key: "axisGridColor"
            }
        }
    }, res;
};

exports.getFunnelTheme = getFunnelTheme, exports.funnel = (0, exports.getFunnelTheme)();
