// @visactor/vchart/cjs/theme/builtin/common/series/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.series = void 0;

const area_1 = require("./area"), bar_1 = require("./bar"), circular_progress_1 = require("./circular-progress"), dot_1 = require("./dot"), funnel_1 = require("./funnel"), gauge_1 = require("./gauge"), gauge_pointer_1 = require("./gauge-pointer"), line_1 = require("./line"), linear_progress_1 = require("./linear-progress"), link_1 = require("./link"), map_1 = require("./map"), pie_1 = require("./pie"), radar_1 = require("./radar"), scatter_1 = require("./scatter"), waterfall_1 = require("./waterfall"), word_cloud_1 = require("./word-cloud"), treemap_1 = require("./treemap"), sunburst_1 = require("./sunburst"), rangeColumn_1 = require("./rangeColumn"), circle_packing_1 = require("./circle-packing"), heatmap_1 = require("./heatmap"), sankey_1 = require("./sankey"), rose_1 = require("./rose"), box_plot_1 = require("./box-plot"), correlation_1 = require("./correlation"), liquid_1 = require("./liquid"), venn_1 = require("./venn"), mosaic_1 = require("./mosaic");

exports.series = {
    scatter: scatter_1.scatter,
    line: line_1.line,
    area: area_1.area,
    bar: bar_1.bar,
    pie: pie_1.pie,
    map: map_1.map,
    radar: radar_1.radar,
    dot: dot_1.dot,
    link: link_1.link,
    wordCloud: word_cloud_1.wordCloud,
    funnel: funnel_1.funnel,
    linearProgress: linear_progress_1.linearProgress,
    circularProgress: circular_progress_1.circularProgress,
    waterfall: waterfall_1.waterfall,
    gauge: gauge_1.gauge,
    gaugePointer: gauge_pointer_1.gaugePointer,
    treemap: treemap_1.treemap,
    sunburst: sunburst_1.sunburst,
    rangeColumn: rangeColumn_1.rangeColumn,
    circlePacking: circle_packing_1.circlePacking,
    heatmap: heatmap_1.heatmap,
    sankey: sankey_1.sankey,
    rose: rose_1.rose,
    boxPlot: box_plot_1.boxPlot,
    correlation: correlation_1.correlation,
    liquid: liquid_1.liquid,
    venn: venn_1.venn,
    mosaic: mosaic_1.mosaic
};
