// @visactor/vrender-animate/cjs/custom/register.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.MoveRotateOut = exports.MoveScaleOut = exports.SpinOut = exports.GrowOut = exports.SlideOut = exports.MoveRotateIn = exports.MoveScaleIn = exports.SpinIn = exports.GrowIn = exports.SlideIn = exports.SlideOutRichText = exports.SlideRichText = exports.OutputRichText = exports.InputRichText = exports.InputText = exports.LabelItemDisappear = exports.LabelItemAppear = exports.MotionPath = exports.Update = exports.State = exports.RotateOut = exports.RotateIn = exports.MoveOut = exports.MoveIn = exports.ScaleOut = exports.ScaleIn = exports.PoptipDisappear = exports.PoptipAppear = exports.IncreaseCount = exports.GrowWidthOut = exports.GrowWidthIn = exports.GrowRadiusOut = exports.GrowRadiusIn = exports.GrowPointsYOut = exports.GrowPointsYIn = exports.GrowPointsXOut = exports.GrowPointsXIn = exports.GrowPointsOut = exports.GrowPointsIn = exports.GrowHeightOut = exports.GrowHeightIn = exports.GrowCenterOut = exports.GrowCenterIn = exports.GrowAngleOut = exports.GrowAngleIn = exports.FadeOut = exports.FadeIn = exports.ClipOut = exports.ClipIn = exports.registerCustomAnimate = void 0, 
exports.Pixelation = exports.GaussianBlur = exports.Glitch = exports.Particle = exports.Distortion = exports.Grayscale = exports.Dissolve = exports.StreamLight = exports.FromTo = exports.GroupFadeOut = exports.GroupFadeIn = exports.PulseAnimate = exports.StrokeOut = exports.StrokeIn = void 0;

const animate_executor_1 = require("../executor/animate-executor"), clip_1 = require("./clip");

Object.defineProperty(exports, "ClipIn", {
    enumerable: !0,
    get: function() {
        return clip_1.ClipIn;
    }
}), Object.defineProperty(exports, "ClipOut", {
    enumerable: !0,
    get: function() {
        return clip_1.ClipOut;
    }
});

const fade_1 = require("./fade");

Object.defineProperty(exports, "FadeIn", {
    enumerable: !0,
    get: function() {
        return fade_1.FadeIn;
    }
}), Object.defineProperty(exports, "FadeOut", {
    enumerable: !0,
    get: function() {
        return fade_1.FadeOut;
    }
});

const growAngle_1 = require("./growAngle");

Object.defineProperty(exports, "GrowAngleIn", {
    enumerable: !0,
    get: function() {
        return growAngle_1.GrowAngleIn;
    }
}), Object.defineProperty(exports, "GrowAngleOut", {
    enumerable: !0,
    get: function() {
        return growAngle_1.GrowAngleOut;
    }
});

const growCenter_1 = require("./growCenter");

Object.defineProperty(exports, "GrowCenterIn", {
    enumerable: !0,
    get: function() {
        return growCenter_1.GrowCenterIn;
    }
}), Object.defineProperty(exports, "GrowCenterOut", {
    enumerable: !0,
    get: function() {
        return growCenter_1.GrowCenterOut;
    }
});

const growHeight_1 = require("./growHeight");

Object.defineProperty(exports, "GrowHeightIn", {
    enumerable: !0,
    get: function() {
        return growHeight_1.GrowHeightIn;
    }
}), Object.defineProperty(exports, "GrowHeightOut", {
    enumerable: !0,
    get: function() {
        return growHeight_1.GrowHeightOut;
    }
});

const growPoints_1 = require("./growPoints");

Object.defineProperty(exports, "GrowPointsIn", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsIn;
    }
}), Object.defineProperty(exports, "GrowPointsOut", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsOut;
    }
}), Object.defineProperty(exports, "GrowPointsXIn", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsXIn;
    }
}), Object.defineProperty(exports, "GrowPointsXOut", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsXOut;
    }
}), Object.defineProperty(exports, "GrowPointsYIn", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsYIn;
    }
}), Object.defineProperty(exports, "GrowPointsYOut", {
    enumerable: !0,
    get: function() {
        return growPoints_1.GrowPointsYOut;
    }
});

const growRadius_1 = require("./growRadius");

Object.defineProperty(exports, "GrowRadiusIn", {
    enumerable: !0,
    get: function() {
        return growRadius_1.GrowRadiusIn;
    }
}), Object.defineProperty(exports, "GrowRadiusOut", {
    enumerable: !0,
    get: function() {
        return growRadius_1.GrowRadiusOut;
    }
});

const growWidth_1 = require("./growWidth");

Object.defineProperty(exports, "GrowWidthIn", {
    enumerable: !0,
    get: function() {
        return growWidth_1.GrowWidthIn;
    }
}), Object.defineProperty(exports, "GrowWidthOut", {
    enumerable: !0,
    get: function() {
        return growWidth_1.GrowWidthOut;
    }
});

const input_text_1 = require("./input-text");

Object.defineProperty(exports, "InputText", {
    enumerable: !0,
    get: function() {
        return input_text_1.InputText;
    }
});

const label_item_animate_1 = require("./label-item-animate");

Object.defineProperty(exports, "LabelItemAppear", {
    enumerable: !0,
    get: function() {
        return label_item_animate_1.LabelItemAppear;
    }
}), Object.defineProperty(exports, "LabelItemDisappear", {
    enumerable: !0,
    get: function() {
        return label_item_animate_1.LabelItemDisappear;
    }
});

const number_1 = require("./number");

Object.defineProperty(exports, "IncreaseCount", {
    enumerable: !0,
    get: function() {
        return number_1.IncreaseCount;
    }
});

const poptip_animate_1 = require("./poptip-animate");

Object.defineProperty(exports, "PoptipAppear", {
    enumerable: !0,
    get: function() {
        return poptip_animate_1.PoptipAppear;
    }
}), Object.defineProperty(exports, "PoptipDisappear", {
    enumerable: !0,
    get: function() {
        return poptip_animate_1.PoptipDisappear;
    }
});

const input_richtext_1 = require("./richtext/input-richtext");

Object.defineProperty(exports, "InputRichText", {
    enumerable: !0,
    get: function() {
        return input_richtext_1.InputRichText;
    }
});

const output_richtext_1 = require("./richtext/output-richtext");

Object.defineProperty(exports, "OutputRichText", {
    enumerable: !0,
    get: function() {
        return output_richtext_1.OutputRichText;
    }
});

const slide_richtext_1 = require("./richtext/slide-richtext");

Object.defineProperty(exports, "SlideRichText", {
    enumerable: !0,
    get: function() {
        return slide_richtext_1.SlideRichText;
    }
});

const slide_out_richtext_1 = require("./richtext/slide-out-richtext");

Object.defineProperty(exports, "SlideOutRichText", {
    enumerable: !0,
    get: function() {
        return slide_out_richtext_1.SlideOutRichText;
    }
});

const scale_1 = require("./scale");

Object.defineProperty(exports, "ScaleIn", {
    enumerable: !0,
    get: function() {
        return scale_1.ScaleIn;
    }
}), Object.defineProperty(exports, "ScaleOut", {
    enumerable: !0,
    get: function() {
        return scale_1.ScaleOut;
    }
});

const state_1 = require("./state");

Object.defineProperty(exports, "State", {
    enumerable: !0,
    get: function() {
        return state_1.State;
    }
});

const story_1 = require("./story");

Object.defineProperty(exports, "GrowIn", {
    enumerable: !0,
    get: function() {
        return story_1.GrowIn;
    }
}), Object.defineProperty(exports, "GrowOut", {
    enumerable: !0,
    get: function() {
        return story_1.GrowOut;
    }
}), Object.defineProperty(exports, "MoveRotateIn", {
    enumerable: !0,
    get: function() {
        return story_1.MoveRotateIn;
    }
}), Object.defineProperty(exports, "MoveRotateOut", {
    enumerable: !0,
    get: function() {
        return story_1.MoveRotateOut;
    }
}), Object.defineProperty(exports, "MoveScaleIn", {
    enumerable: !0,
    get: function() {
        return story_1.MoveScaleIn;
    }
}), Object.defineProperty(exports, "MoveScaleOut", {
    enumerable: !0,
    get: function() {
        return story_1.MoveScaleOut;
    }
}), Object.defineProperty(exports, "PulseAnimate", {
    enumerable: !0,
    get: function() {
        return story_1.PulseAnimate;
    }
}), Object.defineProperty(exports, "SlideIn", {
    enumerable: !0,
    get: function() {
        return story_1.SlideIn;
    }
}), Object.defineProperty(exports, "SlideOut", {
    enumerable: !0,
    get: function() {
        return story_1.SlideOut;
    }
}), Object.defineProperty(exports, "SpinIn", {
    enumerable: !0,
    get: function() {
        return story_1.SpinIn;
    }
}), Object.defineProperty(exports, "SpinOut", {
    enumerable: !0,
    get: function() {
        return story_1.SpinOut;
    }
}), Object.defineProperty(exports, "StrokeIn", {
    enumerable: !0,
    get: function() {
        return story_1.StrokeIn;
    }
}), Object.defineProperty(exports, "StrokeOut", {
    enumerable: !0,
    get: function() {
        return story_1.StrokeOut;
    }
});

const update_1 = require("./update");

Object.defineProperty(exports, "Update", {
    enumerable: !0,
    get: function() {
        return update_1.Update;
    }
});

const move_1 = require("./move");

Object.defineProperty(exports, "MoveIn", {
    enumerable: !0,
    get: function() {
        return move_1.MoveIn;
    }
}), Object.defineProperty(exports, "MoveOut", {
    enumerable: !0,
    get: function() {
        return move_1.MoveOut;
    }
});

const rotate_1 = require("./rotate");

Object.defineProperty(exports, "RotateIn", {
    enumerable: !0,
    get: function() {
        return rotate_1.RotateIn;
    }
}), Object.defineProperty(exports, "RotateOut", {
    enumerable: !0,
    get: function() {
        return rotate_1.RotateOut;
    }
});

const motionPath_1 = require("./motionPath");

Object.defineProperty(exports, "MotionPath", {
    enumerable: !0,
    get: function() {
        return motionPath_1.MotionPath;
    }
});

const fromTo_1 = require("./fromTo");

Object.defineProperty(exports, "FromTo", {
    enumerable: !0,
    get: function() {
        return fromTo_1.FromTo;
    }
});

const groupFade_1 = require("./groupFade");

Object.defineProperty(exports, "GroupFadeIn", {
    enumerable: !0,
    get: function() {
        return groupFade_1.GroupFadeIn;
    }
}), Object.defineProperty(exports, "GroupFadeOut", {
    enumerable: !0,
    get: function() {
        return groupFade_1.GroupFadeOut;
    }
});

const streamLight_1 = require("./streamLight");

Object.defineProperty(exports, "StreamLight", {
    enumerable: !0,
    get: function() {
        return streamLight_1.StreamLight;
    }
});

const dissolve_1 = require("./disappear/dissolve");

Object.defineProperty(exports, "Dissolve", {
    enumerable: !0,
    get: function() {
        return dissolve_1.Dissolve;
    }
});

const grayscale_1 = require("./disappear/grayscale");

Object.defineProperty(exports, "Grayscale", {
    enumerable: !0,
    get: function() {
        return grayscale_1.Grayscale;
    }
});

const distortion_1 = require("./disappear/distortion");

Object.defineProperty(exports, "Distortion", {
    enumerable: !0,
    get: function() {
        return distortion_1.Distortion;
    }
});

const particle_1 = require("./disappear/particle");

Object.defineProperty(exports, "Particle", {
    enumerable: !0,
    get: function() {
        return particle_1.Particle;
    }
});

const glitch_1 = require("./disappear/glitch");

Object.defineProperty(exports, "Glitch", {
    enumerable: !0,
    get: function() {
        return glitch_1.Glitch;
    }
});

const gaussian_blur_1 = require("./disappear/gaussian-blur");

Object.defineProperty(exports, "GaussianBlur", {
    enumerable: !0,
    get: function() {
        return gaussian_blur_1.GaussianBlur;
    }
});

const pixelation_1 = require("./disappear/pixelation");

Object.defineProperty(exports, "Pixelation", {
    enumerable: !0,
    get: function() {
        return pixelation_1.Pixelation;
    }
});

const registerCustomAnimate = () => {
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("increaseCount", number_1.IncreaseCount), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("fromTo", fromTo_1.FromTo), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("scaleIn", scale_1.ScaleIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("scaleOut", scale_1.ScaleOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growHeightIn", growHeight_1.GrowHeightIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growHeightOut", growHeight_1.GrowHeightOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growWidthIn", growWidth_1.GrowWidthIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growWidthOut", growWidth_1.GrowWidthOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growCenterIn", growCenter_1.GrowCenterIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growCenterOut", growCenter_1.GrowCenterOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("clipIn", clip_1.ClipIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("clipOut", clip_1.ClipOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("fadeIn", fade_1.FadeIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("fadeOut", fade_1.FadeOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsIn", growPoints_1.GrowPointsIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsOut", growPoints_1.GrowPointsOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsXIn", growPoints_1.GrowPointsXIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsXOut", growPoints_1.GrowPointsXOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsYIn", growPoints_1.GrowPointsYIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growPointsYOut", growPoints_1.GrowPointsYOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growAngleIn", growAngle_1.GrowAngleIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growAngleOut", growAngle_1.GrowAngleOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growRadiusIn", growRadius_1.GrowRadiusIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growRadiusOut", growRadius_1.GrowRadiusOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveIn", move_1.MoveIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveOut", move_1.MoveOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("rotateIn", rotate_1.RotateIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("rotateOut", rotate_1.RotateOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("update", update_1.Update), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("state", state_1.State), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("labelItemAppear", label_item_animate_1.LabelItemAppear), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("labelItemDisappear", label_item_animate_1.LabelItemDisappear), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("poptipAppear", poptip_animate_1.PoptipAppear), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("poptipDisappear", poptip_animate_1.PoptipDisappear), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("inputText", input_text_1.InputText), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("inputRichText", input_richtext_1.InputRichText), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("outputRichText", output_richtext_1.OutputRichText), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("slideRichText", slide_richtext_1.SlideRichText), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("slideOutRichText", slide_out_richtext_1.SlideOutRichText), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("slideIn", story_1.SlideIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growIn", story_1.GrowIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("spinIn", story_1.SpinIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveScaleIn", story_1.MoveScaleIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveRotateIn", story_1.MoveRotateIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("strokeIn", story_1.StrokeIn), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("slideOut", story_1.SlideOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("growOut", story_1.GrowOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("spinOut", story_1.SpinOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveScaleOut", story_1.MoveScaleOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("moveRotateOut", story_1.MoveRotateOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("strokeOut", story_1.StrokeOut), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("pulse", story_1.PulseAnimate), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("MotionPath", motionPath_1.MotionPath), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("streamLight", streamLight_1.StreamLight), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("dissolve", dissolve_1.Dissolve), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("grayscale", grayscale_1.Grayscale), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("distortion", distortion_1.Distortion), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("particle", particle_1.Particle), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("glitch", glitch_1.Glitch), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("gaussianBlur", gaussian_blur_1.GaussianBlur), 
    animate_executor_1.AnimateExecutor.registerBuiltInAnimate("pixelation", pixelation_1.Pixelation);
};

exports.registerCustomAnimate = registerCustomAnimate;
