// @visactor/vchart-extension/cjs/charts/candlestick/mark/candlestick.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerCandlestickMark = exports.CandlestickMark = exports.CANDLESTICK_MARK_TYPE = void 0;

const vrender_kits_1 = require("@visactor/vrender-kits"), vchart_1 = require("@visactor/vchart"), vrender_core_1 = require("@visactor/vrender-core"), vchart_2 = require("@visactor/vchart");

exports.CANDLESTICK_MARK_TYPE = "candlestick";

class CandlestickMark extends vchart_1.GlyphMark {
    constructor() {
        super(...arguments), this.type = CandlestickMark.type;
    }
    setGlyphConfig(cfg) {
        super.setGlyphConfig(cfg), this._subMarks = {
            line: {
                type: "line",
                defaultAttributes: {
                    x: 0,
                    y: 0
                }
            },
            box: {
                type: "rect"
            }
        }, this._positionChannels = [ "x", "boxWidth", "open", "close", "high", "low" ], 
        this._channelEncoder = null, this._positionEncoder = (glyphAttrs, datum, g) => {
            const {x: x = g.attribute.x, boxWidth: boxWidth = g.attribute.boxWidth, open: open = g.attribute.open, close: close = g.attribute.close, low: low = g.attribute.low, high: high = g.attribute.high} = glyphAttrs, attributes = {};
            return attributes.line = {
                points: [ {
                    x: x,
                    y: low
                }, {
                    x: x,
                    y: high
                } ]
            }, attributes.box = {
                x: x - boxWidth / 2,
                x1: x + boxWidth / 2,
                y: Math.min(open, close),
                y1: Math.max(open, close),
                drawStrokeWhenZeroWH: !0
            }, attributes;
        };
    }
    _getDefaultStyle() {
        return Object.assign({}, super._getDefaultStyle());
    }
}

exports.CandlestickMark = CandlestickMark, CandlestickMark.type = exports.CANDLESTICK_MARK_TYPE;

const registerCandlestickMark = () => {
    (0, vchart_1.registerGlyphMark)(), (0, vrender_kits_1.registerLine)(), (0, vrender_kits_1.registerRect)(), 
    vchart_2.Factory.registerGraphicComponent("line", (attrs => (0, vrender_core_1.createLine)(attrs))), 
    vchart_2.Factory.registerGraphicComponent("rect", (attrs => (0, vrender_core_1.createRect)(attrs))), 
    vchart_2.Factory.registerMark(CandlestickMark.type, CandlestickMark);
};

exports.registerCandlestickMark = registerCandlestickMark;
