// @visactor/vchart/cjs/plugin/components/tooltip-handler/utils/position.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.isFixedTooltipPositionPattern = exports.isGlobalTooltipPositionPattern = exports.getPositionType = exports.positionType = exports.getActualTooltipPositionValue = void 0;

const vutils_1 = require("@visactor/vutils"), getActualTooltipPositionValue = (position, event) => {
    let result;
    if ((0, vutils_1.isValid)(position)) if ((0, vutils_1.isNumber)(position)) result = position; else if ((0, 
    vutils_1.isFunction)(position)) {
        const tooltipPosition = position(event);
        (0, vutils_1.isNumber)(tooltipPosition) && (result = tooltipPosition);
    }
    return result;
};

exports.getActualTooltipPositionValue = getActualTooltipPositionValue, exports.positionType = {
    left: [ -2, 0 ],
    right: [ 2, 0 ],
    top: [ 0, -2 ],
    bottom: [ 0, 2 ],
    insideTop: [ 0, -1.5 ],
    insideBottom: [ 0, 1.5 ],
    insideLeft: [ -1.5, 0 ],
    insideRight: [ 1.5, 0 ],
    insideTopLeft: [ -1.5, -1.5 ],
    insideTopRight: [ 1.5, -1.5 ],
    insideBottomLeft: [ -1.5, 1.5 ],
    insideBottomRight: [ 1.5, 1.5 ],
    lt: [ -2, -2 ],
    tl: [ -2, -2 ],
    rt: [ 2, -2 ],
    tr: [ 2, -2 ],
    bl: [ -2, 2 ],
    lb: [ -2, 2 ],
    br: [ 2, 2 ],
    rb: [ 2, 2 ],
    inside: [ 0, 0 ],
    center: [ 0, 0 ],
    centerBottom: [ 0, 1 ],
    centerTop: [ 0, -1 ],
    centerLeft: [ -1, 0 ],
    centerRight: [ 1, 0 ]
};

const getPositionType = (position, dim, defaultCase = 2) => {
    var _a, _b;
    return null !== (_b = null === (_a = exports.positionType[position]) || void 0 === _a ? void 0 : _a["x" === dim ? 0 : 1]) && void 0 !== _b ? _b : defaultCase;
};

exports.getPositionType = getPositionType;

const isGlobalTooltipPositionPattern = obj => (0, vutils_1.isObject)(obj) && ((0, 
vutils_1.isValid)(obj.left) || (0, vutils_1.isValid)(obj.right) || (0, vutils_1.isValid)(obj.top) || (0, 
vutils_1.isValid)(obj.bottom));

exports.isGlobalTooltipPositionPattern = isGlobalTooltipPositionPattern;

const isFixedTooltipPositionPattern = obj => (0, vutils_1.isObject)(obj) && ((0, 
vutils_1.isValid)(obj.x) || (0, vutils_1.isValid)(obj.y));

exports.isFixedTooltipPositionPattern = isFixedTooltipPositionPattern;
