// @visactor/vchart-extension/esm/charts/sequence-scatter-kde/sequence-scatter-kde.js
import { VChart, BaseChart, ScatterChart } from "@visactor/vchart";

import { SequenceScatterKDEChartSpecTransformer } from "./sequence-scatter-kde-transformer";

export class SequenceScatterKDE extends BaseChart {
    constructor() {
        super(...arguments), this.type = "sequenceScatterKDE", this.transformerConstructor = SequenceScatterKDEChartSpecTransformer;
    }
    init() {
        this.isValid() && super.init();
    }
    isValid() {
        var _a, _b, _c, _d;
        const {xField: xField, yField: yField, data: data} = this._spec;
        return xField && yField ? !!data || (null === (_d = (_c = this._option).onError) || void 0 === _d || _d.call(_c, "Data is required"), 
        !1) : (null === (_b = (_a = this._option).onError) || void 0 === _b || _b.call(_a, "Missing Required Config: `xField`, `yField` "), 
        !1);
    }
}

SequenceScatterKDE.type = "sequenceScatterKDE", SequenceScatterKDE.view = "singleDefault", 
SequenceScatterKDE.transformerConstructor = SequenceScatterKDEChartSpecTransformer;

export const registerSequenceScatterKDE = option => {
    const vchartConstructor = (null == option ? void 0 : option.VChart) || VChart;
    vchartConstructor && vchartConstructor.useChart([ SequenceScatterKDE, ScatterChart ]);
};
