// @visactor/vchart-extension/esm/charts/axis-3d/util.js
import { getCombinedSizeOfRegions } from "@visactor/vchart";

export const getUpdateAttributeOfZAxis = (axis, ignoreGrid) => {
    var _a;
    const regionSize = getCombinedSizeOfRegions(axis.getRegions()), {width: width} = axis.getLayoutRect(), axisLength = width, depth = axis.layout3dBox ? axis.layout3dBox.length : 0, depthZ = axis.layout3dBox ? axis.layout3dBox.width : 0, end = {
        x: depth,
        y: 0
    }, items = axis.getLabelItems(axisLength), attrs = {
        start: {
            x: 0,
            y: 0
        },
        end: end,
        title: {
            text: axis._spec.title.text || axis._dataFieldText,
            maxWidth: axis._getTitleLimit(!1)
        },
        items: items,
        scale: axis._scale.clone()
    }, directionStr = null !== (_a = axis.directionStr) && void 0 !== _a ? _a : "r2l";
    let anchor3d = [ 0, 0 ], alpha = -Math.PI / 2, z = 0;
    return "l2r" === directionStr && (z = axis.layout3dBox.length, anchor3d = [ 0, 0, 0 ], 
    alpha = Math.PI / 2), attrs.z = z, attrs.alpha = alpha, attrs.anchor3d = anchor3d, 
    ignoreGrid || (attrs.grid = {
        type: "line",
        start: {
            x: 0,
            y: 0
        },
        end: end,
        items: items[0],
        verticalFactor: axis._axisStyle.verticalFactor,
        depth: depthZ,
        length: ignoreGrid ? 0 : regionSize.height,
        alpha: alpha,
        z: z,
        anchor3d: anchor3d
    }), attrs;
};
