// @visactor/vchart-extension/cjs/charts/pie-3d/series.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerPie3dSeries = exports.Pie3dSeries = void 0;

const arc_3d_1 = require("../3d/arc-3d"), constant_1 = require("./constant"), enum_1 = require("../3d/enum"), vchart_1 = require("@visactor/vchart"), series_spec_transformer_1 = require("./series-spec-transformer"), animation_1 = require("./animation"), theme_1 = require("./theme");

class Pie3dSeries extends vchart_1.BasePieSeries {
    constructor() {
        super(...arguments), this.type = enum_1.SeriesType3dEnum.pie3d, this._pieMarkName = "pie3d", 
        this._pieMarkType = "arc3d", this.transformerConstructor = series_spec_transformer_1.Pie3dSeriesSpecTransformer;
    }
    setAttrFromSpec() {
        var _a, _b;
        super.setAttrFromSpec(), this._angle3d = null !== (_b = null === (_a = this._spec) || void 0 === _a ? void 0 : _a.angle3d) && void 0 !== _b ? _b : -Math.PI / 3;
    }
    initMarkStyle() {
        super.initMarkStyle();
        const pieMark = this._pieMark;
        pieMark && this.setMarkStyle(pieMark, {
            beta: () => this._angle3d
        }, "normal", vchart_1.AttributeLevel.Series);
    }
    initLabelMarkStyle(textMark, spec = {}) {
        textMark && (super.initLabelMarkStyle(textMark), this.setMarkStyle(textMark, {
            support3d: !0
        }, void 0, vchart_1.AttributeLevel.Mark));
    }
}

exports.Pie3dSeries = Pie3dSeries, Pie3dSeries.type = enum_1.SeriesType3dEnum.pie3d, 
Pie3dSeries.mark = constant_1.pie3dSeriesMark, Pie3dSeries.builtInTheme = {
    pie3d: theme_1.pie3d
}, Pie3dSeries.transformerConstructor = series_spec_transformer_1.Pie3dSeriesSpecTransformer;

const registerPie3dSeries = () => {
    (0, animation_1.registerPie3dAnimation)(), (0, arc_3d_1.registerArc3dMark)(), vchart_1.Factory.registerSeries(Pie3dSeries.type, Pie3dSeries);
};

exports.registerPie3dSeries = registerPie3dSeries;
