// @visactor/vchart-extension/cjs/charts/candlestick/series/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.CandlestickSeriesMark = exports.CandlestickMarkNameEnum = exports.CANDLESTICK_TOOLTIP_KEYS = exports.CANDLESTICK_SERIES_TYPE = exports.CANDLESTICK_CHART_TYPE = void 0;

const vchart_1 = require("@visactor/vchart");

var CANDLESTICK_TOOLTIP_KEYS, CandlestickMarkNameEnum;

exports.CANDLESTICK_CHART_TYPE = "candlestick", exports.CANDLESTICK_SERIES_TYPE = "candlestick", 
function(CANDLESTICK_TOOLTIP_KEYS) {
    CANDLESTICK_TOOLTIP_KEYS.OPEN = "open", CANDLESTICK_TOOLTIP_KEYS.HIGH = "high", 
    CANDLESTICK_TOOLTIP_KEYS.LOW = "low", CANDLESTICK_TOOLTIP_KEYS.CLOSE = "close", 
    CANDLESTICK_TOOLTIP_KEYS.SERIES_FIELD = "seriesField";
}(CANDLESTICK_TOOLTIP_KEYS = exports.CANDLESTICK_TOOLTIP_KEYS || (exports.CANDLESTICK_TOOLTIP_KEYS = {})), 
function(CandlestickMarkNameEnum) {
    CandlestickMarkNameEnum.candlestick = "candlestick";
}(CandlestickMarkNameEnum = exports.CandlestickMarkNameEnum || (exports.CandlestickMarkNameEnum = {})), 
exports.CandlestickSeriesMark = Object.assign(Object.assign({}, vchart_1.baseSeriesMark), {
    candlestick: {
        name: "candlestick",
        type: "candlestick"
    }
});
