// @visactor/vchart-extension/cjs/charts/sequence-scatter-pixel/sequence-scatter-pixel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerSequenceScatterPixel = exports.SequenceScatterPixel = void 0;

const vchart_1 = require("@visactor/vchart"), sequence_scatter_pixel_transformer_1 = require("./sequence-scatter-pixel-transformer");

class SequenceScatterPixel extends vchart_1.BaseChart {
    constructor() {
        super(...arguments), this.type = "sequenceScatterPixel", this.transformerConstructor = sequence_scatter_pixel_transformer_1.SequenceScatterPixelChartSpecTransformer;
    }
    init() {
        this.isValid() && super.init();
    }
    isValid() {
        var _a, _b, _c, _d;
        const {xField: xField, yField: yField, data: data} = this._spec;
        return xField && yField ? !!data || (null === (_d = (_c = this._option).onError) || void 0 === _d || _d.call(_c, "Data is required"), 
        !1) : (null === (_b = (_a = this._option).onError) || void 0 === _b || _b.call(_a, "Missing Required Config: `xField`, `yField` "), 
        !1);
    }
}

exports.SequenceScatterPixel = SequenceScatterPixel, SequenceScatterPixel.type = "sequenceScatterPixel", 
SequenceScatterPixel.view = "singleDefault", SequenceScatterPixel.transformerConstructor = sequence_scatter_pixel_transformer_1.SequenceScatterPixelChartSpecTransformer;

const registerSequenceScatterPixel = option => {
    const vchartConstructor = (null == option ? void 0 : option.VChart) || vchart_1.VChart;
    vchartConstructor && vchartConstructor.useChart([ SequenceScatterPixel, vchart_1.ScatterChart ]);
};

exports.registerSequenceScatterPixel = registerSequenceScatterPixel;
