// @visactor/vchart-extension/esm/charts/sequence-scatter-link/sequence-scatter-link.js
import { VChart, BaseChart, ScatterChart } from "@visactor/vchart";

import { SequenceScatterLinkChartSpecTransformer } from "./sequence-scatter-link-transformer";

export class SequenceScatterLink extends BaseChart {
    constructor() {
        super(...arguments), this.type = "sequenceScatterLink", this.transformerConstructor = SequenceScatterLinkChartSpecTransformer;
    }
    init() {
        this.isValid() && super.init();
    }
    isValid() {
        var _a, _b, _c, _d;
        const {xField: xField, yField: yField, data: data} = this._spec;
        return xField && yField ? !!data || (null === (_d = (_c = this._option).onError) || void 0 === _d || _d.call(_c, "Data is required"), 
        !1) : (null === (_b = (_a = this._option).onError) || void 0 === _b || _b.call(_a, "Missing Required Config: `xField`, `yField` "), 
        !1);
    }
}

SequenceScatterLink.type = "sequenceScatterLink", SequenceScatterLink.view = "singleDefault", 
SequenceScatterLink.transformerConstructor = SequenceScatterLinkChartSpecTransformer;

export const registerSequenceScatterLink = option => {
    const vchartConstructor = (null == option ? void 0 : option.VChart) || VChart;
    vchartConstructor && vchartConstructor.useChart([ SequenceScatterLink, ScatterChart ]);
};
