// @visactor/vchart-extension/esm/charts/word-cloud-3d/chart.js
import { BaseWordCloudChart, Factory } from "@visactor/vchart";

import { register3DPlugin } from "../3d/plugin";

import { registerWordCloud3dSeries, registerWordCloudShape3dSeries } from "./series";

import { WordCloud3dChartSpecTransformer } from "./chart-spec-transformer";

import { SeriesType3dEnum } from "../3d/enum";

export class WordCloud3dChart extends BaseWordCloudChart {
    constructor() {
        super(...arguments), this.transformerConstructor = WordCloud3dChartSpecTransformer, 
        this.type = "wordCloud3d", this.seriesType = SeriesType3dEnum.wordCloud3d;
    }
}

WordCloud3dChart.type = "wordCloud3d", WordCloud3dChart.seriesType = SeriesType3dEnum.wordCloud3d, 
WordCloud3dChart.transformerConstructor = WordCloud3dChartSpecTransformer;

export const registerWordCloud3dChart = () => {
    register3DPlugin(), registerWordCloud3dSeries(), Factory.registerChart(WordCloud3dChart.type, WordCloud3dChart);
};

export const registerWordCloudShape3dChart = () => {
    register3DPlugin(), registerWordCloudShape3dSeries(), register3DPlugin(), registerWordCloud3dSeries(), 
    Factory.registerChart(WordCloud3dChart.type, WordCloud3dChart);
};
