// @visactor/vchart/esm/mark/index.js
import { LineMark, registerLineMark } from "./line";

import { SymbolMark, registerSymbolMark } from "./symbol";

import { GroupMark, registerGroupMark } from "./group";

import { RuleMark, registerRuleMark } from "./rule";

import { TextMark, registerTextMark } from "./text";

import { AreaMark, registerAreaMark } from "./area";

import { RectMark, registerRectMark } from "./rect";

import { PathMark, registerPathMark } from "./path";

import { ArcMark, BaseArcMark, registerArcMark } from "./arc";

import { ComponentMark, registerComponentMark } from "./component";

import { RippleMark, registerRippleMark } from "./ripple";

import { GlyphMark, registerGlyphMark } from "./glyph";

import { BaseMark } from "./base";

import { PolygonMark, registerPolygonMark } from "./polygon/polygon";

import { ImageMark, registerImageMark } from "./image";

import { registerMarkFilterTransform } from "./transform/filter";

import { registerMarkMapTransform } from "./transform/map";

import { BasePolygonMark } from "./polygon/base-polygon";

import { MarkTypeEnum } from "./interface/type";

export { MarkTypeEnum, LineMark, SymbolMark, GroupMark, RuleMark, TextMark, AreaMark, RectMark, PathMark, GlyphMark, BaseArcMark, ArcMark, ComponentMark, BasePolygonMark, PolygonMark, RippleMark, ImageMark, BaseMark };

export { registerLineMark, registerSymbolMark, registerGroupMark, registerRuleMark, registerTextMark, registerAreaMark, registerRectMark, registerPathMark, registerArcMark, registerPolygonMark, registerGlyphMark, registerRippleMark, registerImageMark, registerComponentMark, registerMarkMapTransform, registerMarkFilterTransform };

export const registerAllMarks = () => {
    registerGroupMark(), registerLineMark(), registerSymbolMark(), registerRuleMark(), 
    registerTextMark(), registerAreaMark(), registerRectMark(), registerPathMark(), 
    registerArcMark(), registerPolygonMark(), registerImageMark(), registerRippleMark(), 
    registerComponentMark();
};
