// @visactor/vchart/cjs/core/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.measureText = exports.warn = exports.lookup = exports.registerDataSetInstanceTransform = exports.registerDataSetInstanceParser = exports.getSpecInfo = exports.transformToGraphic = exports.StreamLight = exports.ManualTicker = exports.createSymbol = exports.createLine = exports.createRect = exports.createArea = exports.createArc = exports.createText = exports.createRichText = exports.createGroup = exports.vglobal = exports.ChartEvent = exports.STATE_VALUE_ENUM = exports.TransformLevel = exports.AttributeLevel = exports.version = exports.Factory = exports.VChart = void 0;

const vchart_1 = require("./vchart");

Object.defineProperty(exports, "VChart", {
    enumerable: !0,
    get: function() {
        return vchart_1.VChart;
    }
});

const factory_1 = require("./factory");

Object.defineProperty(exports, "Factory", {
    enumerable: !0,
    get: function() {
        return factory_1.Factory;
    }
}), exports.version = "2.0.4", __exportStar(require("../event/interface"), exports), 
__exportStar(require("../theme/interface"), exports), __exportStar(require("../model/interface"), exports), 
__exportStar(require("./interface"), exports), __exportStar(require("../typings"), exports), 
__exportStar(require("../constant/base"), exports), __exportStar(require("../constant/data"), exports), 
__exportStar(require("../constant/layout"), exports);

var attribute_1 = require("../constant/attribute");

Object.defineProperty(exports, "AttributeLevel", {
    enumerable: !0,
    get: function() {
        return attribute_1.AttributeLevel;
    }
});

var initialize_1 = require("../data/initialize");

Object.defineProperty(exports, "TransformLevel", {
    enumerable: !0,
    get: function() {
        return initialize_1.TransformLevel;
    }
});

var interface_1 = require("../compile/mark/interface");

Object.defineProperty(exports, "STATE_VALUE_ENUM", {
    enumerable: !0,
    get: function() {
        return interface_1.STATE_VALUE_ENUM;
    }
});

var event_1 = require("../constant/event");

Object.defineProperty(exports, "ChartEvent", {
    enumerable: !0,
    get: function() {
        return event_1.ChartEvent;
    }
}), __exportStar(require("../typings/spec/index"), exports), __exportStar(require("../typings/visual"), exports), 
__exportStar(require("../typings/tooltip"), exports), __exportStar(require("../theme/index"), exports);

var vrender_core_1 = require("@visactor/vrender-core");

Object.defineProperty(exports, "vglobal", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.vglobal;
    }
}), Object.defineProperty(exports, "createGroup", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createGroup;
    }
}), Object.defineProperty(exports, "createRichText", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createRichText;
    }
}), Object.defineProperty(exports, "createText", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createText;
    }
}), Object.defineProperty(exports, "createArc", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createArc;
    }
}), Object.defineProperty(exports, "createArea", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createArea;
    }
}), Object.defineProperty(exports, "createRect", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createRect;
    }
}), Object.defineProperty(exports, "createLine", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createLine;
    }
}), Object.defineProperty(exports, "createSymbol", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createSymbol;
    }
});

var vrender_animate_1 = require("@visactor/vrender-animate");

Object.defineProperty(exports, "ManualTicker", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.ManualTicker;
    }
}), Object.defineProperty(exports, "StreamLight", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.StreamLight;
    }
}), __exportStar(require("../util/space"), exports);

var style_1 = require("../util/style");

Object.defineProperty(exports, "transformToGraphic", {
    enumerable: !0,
    get: function() {
        return style_1.transformToGraphic;
    }
});

var util_1 = require("../component/util");

Object.defineProperty(exports, "getSpecInfo", {
    enumerable: !0,
    get: function() {
        return util_1.getSpecInfo;
    }
});

var register_1 = require("../data/register");

Object.defineProperty(exports, "registerDataSetInstanceParser", {
    enumerable: !0,
    get: function() {
        return register_1.registerDataSetInstanceParser;
    }
}), Object.defineProperty(exports, "registerDataSetInstanceTransform", {
    enumerable: !0,
    get: function() {
        return register_1.registerDataSetInstanceTransform;
    }
});

var lookup_1 = require("../data/transforms/lookup");

Object.defineProperty(exports, "lookup", {
    enumerable: !0,
    get: function() {
        return lookup_1.lookup;
    }
});

var debug_1 = require("../util/debug");

Object.defineProperty(exports, "warn", {
    enumerable: !0,
    get: function() {
        return debug_1.warn;
    }
});

var text_1 = require("../util/text");

Object.defineProperty(exports, "measureText", {
    enumerable: !0,
    get: function() {
        return text_1.measureText;
    }
});
