// @visactor/react-vchart/cjs/containers/withContainer.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function(o, v) {
    Object.defineProperty(o, "default", {
        enumerable: !0,
        value: v
    });
} : function(o, v) {
    o.default = v;
}), __importStar = this && this.__importStar || function(mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (null != mod) for (var k in mod) "default" !== k && Object.prototype.hasOwnProperty.call(mod, k) && __createBinding(result, mod, k);
    return __setModuleDefault(result, mod), result;
}, __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

const react_1 = __importStar(require("react"));

function withContainer(Comp, name = "ChartContainer", getProps) {
    const Cls = react_1.default.forwardRef(((props, ref) => {
        const container = (0, react_1.useRef)(null), [inited, setInited] = (0, react_1.useState)(!1), {className: className, style: style, width: width, height: height, id: id} = props, options = __rest(props, [ "className", "style", "width", "height", "id" ]);
        return (0, react_1.useLayoutEffect)((() => {
            setInited(!0);
        }), []), react_1.default.createElement("div", {
            ref: container,
            id: id,
            className: className,
            style: Object.assign({
                position: "relative",
                height: height || "100%",
                width: width || "100%"
            }, style)
        }, inited ? react_1.default.createElement(Comp, Object.assign({
            ref: ref,
            container: container.current
        }, getProps ? getProps(options) : options)) : react_1.default.createElement(react_1.default.Fragment, null));
    }));
    return Cls.displayName = name || Comp.name, Cls;
}

exports.default = withContainer;
