// @visactor/vchart-extension/esm/charts/ranking-list/ranking-list.js
import { VChart, BaseChart, BarChart } from "@visactor/vchart";

import { RankingListChartSpecTransformer } from "./ranking-list-transformer";

export class RankingList extends BaseChart {
    constructor() {
        super(...arguments), this.type = "rankingList", this.transformerConstructor = RankingListChartSpecTransformer;
    }
    init() {
        this.isValid() && super.init();
    }
    isValid() {
        var _a, _b, _c, _d;
        const {xField: xField, yField: yField, data: data} = this._spec;
        return xField && yField ? !!data || (null === (_d = (_c = this._option).onError) || void 0 === _d || _d.call(_c, "Data is required"), 
        !1) : (null === (_b = (_a = this._option).onError) || void 0 === _b || _b.call(_a, "Missing Required Config: `xField`, `yField` "), 
        !1);
    }
}

RankingList.type = "rankingList", RankingList.view = "singleDefault", RankingList.transformerConstructor = RankingListChartSpecTransformer;

export const registerRankingList = option => {
    const vchartConstructor = (null == option ? void 0 : option.VChart) || VChart;
    vchartConstructor && vchartConstructor.useChart([ RankingList, BarChart ]);
};
