// @visactor/vrender-kits/es/render/contributions/rough/rough-rect.js
var __decorate = this && this.__decorate || function(decorators, target, key, desc) {
    var d, c = arguments.length, r = c < 3 ? target : null === desc ? desc = Object.getOwnPropertyDescriptor(target, key) : desc;
    if ("object" == typeof Reflect && "function" == typeof Reflect.decorate) r = Reflect.decorate(decorators, target, key, desc); else for (var i = decorators.length - 1; i >= 0; i--) (d = decorators[i]) && (r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r);
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}, __metadata = this && this.__metadata || function(k, v) {
    if ("object" == typeof Reflect && "function" == typeof Reflect.metadata) return Reflect.metadata(k, v);
}, __param = this && this.__param || function(paramIndex, decorator) {
    return function(target, key) {
        decorator(target, key, paramIndex);
    };
};

import { RECT_NUMBER_TYPE, DefaultCanvasRectRender, inject, injectable } from "@visactor/vrender-core";

import { RoughBaseRender } from "./base-render";

let RoughCanvasRectRender = class extends RoughBaseRender {
    constructor(canvasRenderer) {
        super(), this.canvasRenderer = canvasRenderer, this.style = "rough", this.type = "rect", 
        this.numberType = RECT_NUMBER_TYPE;
    }
    draw(rect, renderService, drawContext, params) {
        this.doDraw(rect, renderService, drawContext, params);
    }
};

RoughCanvasRectRender = __decorate([ injectable(), __param(0, inject(DefaultCanvasRectRender)), __metadata("design:paramtypes", [ Object ]) ], RoughCanvasRectRender);

export { RoughCanvasRectRender };
